{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vignette of Pypi-Package hapROH: Calling ROH from an Eigenstrat File\n",
    "This notebook contains an example application of `hapROH`:\n",
    "We will identify ROH using an Eigenstrat file as input. This notebook can serve as a blueprint for your own ROH inference.\n",
    "\n",
    "You will see how to run `hapROH` on single chromosomes, on whole individuals, and sets of individuals, and how to post-process multiple output files into one summary table (including metadata)\n",
    "\n",
    "To learn how to visualize the results, take a look at `plotROH_vignette.ipynb`.\n",
    "\n",
    "@author: Harald Ringbauer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU Count: 128\n"
     ]
    }
   ],
   "source": [
    "### First some Standard Imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "\n",
    "### If you want to use a version other than the default installed hapROH \n",
    "### uncomment the following and set the path to the correct (installed) package\n",
    "#sys.path.insert(0,\"/mnt/archgen/users/hringbauer/git/hapROH/package/\")  # Uncomment to get local package first in path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set the Path\n",
    "You can set the path here to the desired working directory (relative paths below will be resolved relative to this folder). Alternatively, you can give absolute paths in the next steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set path to: /mnt/archgen/users/hringbauer/git/hapROH/Notebooks/Vignettes\n"
     ]
    }
   ],
   "source": [
    "### Fill in your own path here:\n",
    "path = \"/mnt/archgen/users/hringbauer/git/hapROH/Notebooks/Vignettes/\"  # The Path to Package Midway Cluster\n",
    "\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory. Should be HAPSBURG/Notebooks/ParallelRuns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1) Download the data we will use as an example here\n",
    "If you don't already have it, you will need to obtain the target and reference data. Here is how:\n",
    "\n",
    "### Get Target Data\n",
    "We will download an example eigenstrat available via the Reich lab homepage. \n",
    "It contains a handful of individuals from the Chalcolithic Levant (zipped, ~20 MB).\n",
    "You can download it from:\n",
    "`https://reich.hms.harvard.edu/sites/reich.hms.harvard.edu/files/inline-files/Levant_ChL.tar.gz`\n",
    "Download and unpack the files into the desired folder (unpacking will create .geno, .ind, and .snp files).\n",
    "\n",
    "### Get Reference Data\n",
    "You can download the reference data from \n",
    "`https://www.dropbox.com/s/0qhjgo1npeih0bw/1000g1240khdf5.tar.gz?dl=0`\n",
    "\n",
    "This data is key for running hapROH - it is the 1000 Genome data downsampled to bi-allelic 1240k SNPs, from 5008 global reference haplotypes (Note: the file is ~800 MB in size). \n",
    "\n",
    "There is also a metafile datatable that contains information about the reference panel. This table is necessary to run hapROH; do not forget to copy it over as well, in addition to the data for each chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2) Call ROH\n",
    "Now we will run the core part of the package, a wrapper function for the \n",
    "core ROH calling machinery - `callROH_ind()`.\n",
    "\n",
    "We first import it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hapROH.run import callROH_ind"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test calling ROH on a single chromosome from an Individual, with full output printed\n",
    "Set the location of the reference HDF5 and the target eigenstrat path correctly!\n",
    "\n",
    "Input: Reference haplotype file (hdf5, containing genetic map), reference metafile, pseudo-haploid Eigenstrat\n",
    "Output: Output file in folder `folder_out`\n",
    "\n",
    "See example below, you need to set `path_targets` and `h5_path1000G`\n",
    "\n",
    "logfile=False -> All output is printed into sys.out  \n",
    "combine=False -> No individual .csv is created from chromosome .csv files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Minimal Version of `callROH_ind()`\n",
    "A version where the parameters a normal user does not have to set hidden (i.e., set to default values that work well for 1240K data).\n",
    "hapROH is tested for a wide variety of human aDNA data, and the default parameters likely have you covered if you run hapROH on <50 ky old 1240K data.\n",
    "\n",
    "This function will call ROH of individual `iid` in the target eigenstrat given in `path_targets`. We also set the path to the reference HDF5 file (`h5_path1000G`) and the metafile of the reference (`meta_path_ref`), as well as the folder in which to save the output (`folder_out`).\n",
    "\n",
    "Let us run the date for chromosome 20 (`range(20, 21)` selects a list of length 1: `[20]`). We will use one processor (`processes=1`) and print all text output (`output=True`) to the console (`logfile=False`). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I1178\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr20.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 22 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29078\n",
      "Nr of Matching Refs: 29078 / 29078\n",
      "Ref/Alt Matching: 29053 / 29078\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29053 / 29078\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 19815 / 29053\n",
      "Fraction SNPs covered: 0.6820\n",
      "Extraction of 5008 reference haplotypes at 19815 sites complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./ExampleOutput/I1178/chr20/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./ExampleOutput/I1178/chr20/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI1178\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "downsample\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./ExampleData/Levant_ChL\n",
      "h5_path1000g\t/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\n",
      "meta_path_ref\t/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\n",
      "folder_out\t./ExampleOutput/\n",
      "prefix_out\t\n",
      "c\t0.0\n",
      "conPop\t['CEU']\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0.005\n",
      "roh_min_l_initial\t0.02\n",
      "roh_min_l_final\t0.04\n",
      "min_len1\t0.02\n",
      "min_len2\t0.04\n",
      "cutoff_post\t0.999\n",
      "verbose\tTrue\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 159\n",
      "Maximum Gap: 0.4744 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition Matrix:\n",
      "(3, 3)\n",
      "Loaded Observations:\n",
      "(2, 19815)\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 453.517312 mB\n",
      "Total Log likelihood: -5803.614\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./ExampleOutput/I1178/chr20/.\n",
      "Finished calculating posterior, takes 1.208\n",
      "Successfully loaded for PP. from ./ExampleOutput/I1178/chr20/\n",
      "Fraction Markers in ROH: 0.4818\n",
      "Called n=1 ROH Blocks > 4.0 cM\n",
      "Longest Block: 26.10 cM\n",
      "Successfully saved to ./ExampleOutput/I1178/chr20/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "callROH_ind(iid=\"I1178\", chs=range(20, 21), \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\", # The path up to the chr. number\n",
    "          meta_path_ref=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\", \n",
    "          folder_out='./ExampleOutput/',  # Folder where you want to save the results to \n",
    "          processes=1, verbose=True,\n",
    "          readcounts=False, logfile=False, combine=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Power version of `callROH_ind` [does the same as above, but shows more parameters that could be changed]\n",
    "Shows most of the parameters that you can set. \n",
    "\n",
    "Warning: Try to understand what you are doing when changing parameters. For example, the transition parameters are already optimized for 1240k data; changing them can have unintended consequences.\n",
    "\n",
    "All parameters are explained in the docstring of `callROH_ind()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I1178\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr20.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 22 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29078\n",
      "Nr of Matching Refs: 29078 / 29078\n",
      "Ref/Alt Matching: 29053 / 29078\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29053 / 29078\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 19815 / 29053\n",
      "Fraction SNPs covered: 0.6820\n",
      "Extraction of 5008 reference haplotypes at 19815 sites complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./ExampleOutput/I1178/chr20/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./ExampleOutput/I1178/chr20/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI1178\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "downsample\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./ExampleData/Levant_ChL\n",
      "h5_path1000g\t/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\n",
      "meta_path_ref\t/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\n",
      "folder_out\t./ExampleOutput/\n",
      "prefix_out\t\n",
      "c\t0.0\n",
      "conPop\t['CEU']\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0.005\n",
      "roh_min_l_initial\t0.02\n",
      "roh_min_l_final\t0.04\n",
      "min_len1\t0.02\n",
      "min_len2\t0.04\n",
      "cutoff_post\t0.999\n",
      "verbose\tTrue\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 159\n",
      "Maximum Gap: 0.4744 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition Matrix:\n",
      "(3, 3)\n",
      "Loaded Observations:\n",
      "(2, 19815)\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 537.194496 mB\n",
      "Total Log likelihood: -5803.614\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./ExampleOutput/I1178/chr20/.\n",
      "Finished calculating posterior, takes 1.206\n",
      "Successfully loaded for PP. from ./ExampleOutput/I1178/chr20/\n",
      "Fraction Markers in ROH: 0.4818\n",
      "Called n=1 ROH Blocks > 4.0 cM\n",
      "Longest Block: 26.10 cM\n",
      "Successfully saved to ./ExampleOutput/I1178/chr20/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "callROH_ind(iid=\"I1178\", chs=range(20, 21), \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\", # The path up to the chr. number\n",
    "          meta_path_ref=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\", \n",
    "          folder_out='./ExampleOutput/', prefix_out='', \n",
    "          e_model='haploid', p_model='Eigenstrat', post_model='Standard', \n",
    "          low_mem=True, processes=1, delete=False, verbose=True, save=True, save_fp=False, \n",
    "          n_ref=2504, diploid_ref=True, exclude_pops=[], readcounts=False, random_allele=True,\n",
    "          c=0.0, conPop=[\"CEU\"], roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.00, \n",
    "          cutoff_post = 0.999, max_gap=0.005, roh_min_l_initial = 0.02, roh_min_l_final = 0.04,\n",
    "          min_len1 = 0.02, min_len2 = 0.04, logfile=False, combine=False, file_result=\"_roh_full.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example run of whole individual (all chromosomes) with output to logfile\n",
    "Now let us run a whole individual in parallel (for each chromosome), with the output sent to a log file.\n",
    "\n",
    "Attention: `hapROH` requires a non-trivial amount of RAM. For an Eigenstrat, it can spike up to a few GB for a long chromosome with most SNPs covered. Allocate memory accordingly when running multiple processes, or set the number of parallel processes to a lower value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr1/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n"
     ]
    }
   ],
   "source": [
    "callROH_ind(iid=\"I1178\", chs=range(1,23), processes=6, \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\", # The path up to the chr. number\n",
    "          meta_path_ref=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\",\n",
    "          folder_out='./ExampleOutput/', prefix_out='', \n",
    "          e_model=\"haploid\", p_model=\"Eigenstrat\",\n",
    "          random_allele=True, readcounts=False,\n",
    "          delete=False, logfile=True, combine=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run multiple Individuals\n",
    "Now, let's loop over individuals. In practice, I run it as a parallelized sbatch job (setting processes to 1 there) - this way, you can parallelize it with one process per individual and not worry about simultaneous memory spikes. That is a recommended level of parallelization; the individual output files will be combined in a subsequent step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual: I1178\n",
      "Doing Individual I1178...\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr5/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I0644\n",
      "Doing Individual I0644...\n",
      "Set Output Log path: ./ExampleOutput/I0644/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr5/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1160\n",
      "Doing Individual I1160...\n",
      "Set Output Log path: ./ExampleOutput/I1160/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr1/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1152\n",
      "Doing Individual I1152...\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr6/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1168\n",
      "Doing Individual I1168...\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1168/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1168/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1168/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1168/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1168/chr6/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1166\n",
      "Doing Individual I1166...\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1166/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1166/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1166/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1166/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1166/chr4/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1170\n",
      "Doing Individual I1170...\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1170/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1170/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1170/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1170/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1170/chr4/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1165\n",
      "Doing Individual I1165...\n",
      "Set Output Log path: ./ExampleOutput/I1165/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr1/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1182\n",
      "Doing Individual I1182...\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr6/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n"
     ]
    }
   ],
   "source": [
    "### This are all individuals with 400k SNPs covered\n",
    "iids = ['I1178', 'I0644', 'I1160', 'I1152', 'I1168', 'I1166', 'I1170', 'I1165', 'I1182']\n",
    "\n",
    "for iid in iids:\n",
    "    print(f\"Doing Individual: {iid}\")\n",
    "    callROH_ind(iid=iid, chs=range(1,23), processes=6, \n",
    "              path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "              h5_path1000g=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/1240k_binary_chr\", # The path up to the chr. number\n",
    "              meta_path_ref=\"/mnt/archgen/public_data/hapROH_refpanel/1240k_binary/meta_df_all.csv\",\n",
    "              folder_out='./ExampleOutput/', prefix_out='', \n",
    "              e_model=\"haploid\", p_model=\"Eigenstrat\", n_ref=2504,\n",
    "              random_allele=True, readcounts=False,\n",
    "              delete=False, logfile=True, combine=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) Postprocess Results into a single results.csv (optionally, copying in Meta Data)\n",
    "Take individual output .csv files and combine them into one main results .csv file.\n",
    "\n",
    "**Important**: Merging of ROH gaps, and filtering to ROH>x cM happens here. This is a key step when interpreting the output!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from hapsburg.PackagesSupport.pp_individual_roh_csvs import pp_individual_roh\n",
    "from hapROH.run import ppROH_inds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Example Meta File\n",
    "This file will create a minimal meta data file (some plotting functions use individual meta information, such as age),\n",
    "which is then merged into the results when combining individuals.\n",
    "\n",
    "If you have metadata available for your dataset, you can prepare a comma seperated csv table with these headers!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "iids = ['I1178', 'I0644', 'I1160', \n",
    "        'I1152', 'I1168', 'I1166', \n",
    "        'I1170', 'I1165', 'I1182']\n",
    "df = pd.DataFrame({\"iid\":iids})\n",
    "\n",
    "df[\"clst\"] = \"Israel_C\"  # The clst column is necessary - it sets the population label column\n",
    "df[\"age\"] = 5950\n",
    "df[\"lat\"] = 32.974167\n",
    "df[\"lon\"] = 35.331389\n",
    "\n",
    "df.to_csv(\"./ExampleData/meta_blank.csv\", \n",
    "          sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine individual output files into a single ROH output File\n",
    "Combines meta file, individual ROH output files\n",
    "into a summary table, contining statistical information about ROH for each indiviual as a row.\n",
    "\n",
    "This is the table that a lot of the downstream plotting software uses.\n",
    "\n",
    "Creating this table also does additional post-processing of the ROH, in particular merging gaps between ROH. Every gap of length gap [in cM] between two ROH of >min_len1 and >min_len2 is merged.\n",
    "\n",
    "This is by default already done in `hapsb_ind`, but can be also done here. Doing it only in the end is useful e.g. in cases where one wants to try different post-processing and gap-merging: One can set the gap to 0 in hapsb_ind and then here try different values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 9 / 9 Individuals from Meta\n",
      "Saved to: ./ExampleOutput/combined_roh05.csv\n"
     ]
    }
   ],
   "source": [
    "### Postprocess the two Individuals from above and combine into one results .csv\n",
    "iids = ['I1178', 'I0644', 'I1160', 'I1152', \n",
    "        'I1168', 'I1166', 'I1170', 'I1165', 'I1182']\n",
    "\n",
    "df1 = ppROH_inds(iids, meta_path=\"./ExampleData/meta_blank.csv\",  \n",
    "                 base_folder=\"./ExampleOutput/\",\n",
    "                 save_path=\"./ExampleOutput/combined_roh05.csv\", \n",
    "                 verbose=False, min_cm=[4, 8, 12, 20], snp_cm=50, \n",
    "                 gap=0.5, min_len1=2.0, min_len2=4.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Alternative Post-processing]\n",
    "You do not strictly require creating a meta table. You can only provide the list of iids.\n",
    "Note that this will set the population column to a default value for all iids, which affects downstream plotting.\n",
    "You can, of course, manually update this column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "iids = ['I1178', 'I0644', 'I1160', 'I1152', \n",
    "        'I1168', 'I1166', 'I1170', 'I1165', 'I1182']\n",
    "\n",
    "df2 = ppROH_inds(iids, base_folder=\"./ExampleOutput/\", meta_info=False,\n",
    "                 save_path=\"\", meta_path=\"\", verbose=False, min_cm=[4, 8, 12, 20], snp_cm=50, \n",
    "                 gap=0.5, min_len1=2.0, min_len2=4.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the output table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>max_roh</th>\n",
       "      <th>iid</th>\n",
       "      <th>pop</th>\n",
       "      <th>sum_roh&gt;4</th>\n",
       "      <th>n_roh&gt;4</th>\n",
       "      <th>sum_roh&gt;8</th>\n",
       "      <th>n_roh&gt;8</th>\n",
       "      <th>sum_roh&gt;12</th>\n",
       "      <th>n_roh&gt;12</th>\n",
       "      <th>sum_roh&gt;20</th>\n",
       "      <th>n_roh&gt;20</th>\n",
       "      <th>clst</th>\n",
       "      <th>age</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>91.121798</td>\n",
       "      <td>I1178</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>703.292088</td>\n",
       "      <td>30</td>\n",
       "      <td>682.518689</td>\n",
       "      <td>26</td>\n",
       "      <td>625.156411</td>\n",
       "      <td>20</td>\n",
       "      <td>545.074702</td>\n",
       "      <td>15</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9.192801</td>\n",
       "      <td>I1170</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>26.837993</td>\n",
       "      <td>4</td>\n",
       "      <td>18.287802</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.879100</td>\n",
       "      <td>I1166</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>9.593799</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.202509</td>\n",
       "      <td>I1160</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>8.283711</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I0644</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1168</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1152</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1165</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1182</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>5950</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     max_roh    iid       pop   sum_roh>4  n_roh>4   sum_roh>8  n_roh>8  \\\n",
       "0  91.121798  I1178  Israel_C  703.292088       30  682.518689       26   \n",
       "1   9.192801  I1170  Israel_C   26.837993        4   18.287802        2   \n",
       "2   4.879100  I1166  Israel_C    9.593799        2    0.000000        0   \n",
       "3   4.202509  I1160  Israel_C    8.283711        2    0.000000        0   \n",
       "4   0.000000  I0644  Israel_C    0.000000        0    0.000000        0   \n",
       "5   0.000000  I1168  Israel_C    0.000000        0    0.000000        0   \n",
       "6   0.000000  I1152  Israel_C    0.000000        0    0.000000        0   \n",
       "7   0.000000  I1165  Israel_C    0.000000        0    0.000000        0   \n",
       "8   0.000000  I1182  Israel_C    0.000000        0    0.000000        0   \n",
       "\n",
       "   sum_roh>12  n_roh>12  sum_roh>20  n_roh>20      clst   age        lat  \\\n",
       "0  625.156411        20  545.074702        15  Israel_C  5950  32.974167   \n",
       "1    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "2    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "3    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "4    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "5    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "6    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "7    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "8    0.000000         0    0.000000         0  Israel_C  5950  32.974167   \n",
       "\n",
       "         lon  \n",
       "0  35.331389  \n",
       "1  35.331389  \n",
       "2  35.331389  \n",
       "3  35.331389  \n",
       "4  35.331389  \n",
       "5  35.331389  \n",
       "6  35.331389  \n",
       "7  35.331389  \n",
       "8  35.331389  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One individual (I1178) has extremely high amounts of ROH! Well, maybe that's why I chose this dataset as the example :). \n",
    "\n",
    "For more utilities about interpreting and visualizing these ROH results, you can continue in `./plotROH_vignette.ipynb`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
