{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vignette of Pypi-Package hapROH: Calling ROH from an Eigenstrat File\n",
    "This notebook contains an example application of hapROH:\n",
    "We will identify ROH in a target Eigenstrat. Importantly, it can serve as blueprint for your own ROH inference!\n",
    "\n",
    "You will learn how to run hapROH on single chromosomes, on whole individuals, and sets of individuals, and how to post-process multiple output files into one summary datafile (including meta data)\n",
    "\n",
    "If you want to learn how to visualize the results, take a look at the plotting vignette.\n",
    "\n",
    "@author: Harald Ringbauer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU Count: 40\n"
     ]
    }
   ],
   "source": [
    "### Some Code to set right paths on Harald's Machine\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "\n",
    "### If wanting to use specific version and not default installed hapROH version add at head of PATH:\n",
    "#sys.path.insert(0,\"./package/\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set the Path\n",
    "You can set the path here to the path you want to work in (relative data paths will be calculated relative to this folder.) Alternatively you can give absolute paths in the next steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set path to: /mnt/archgen/users/hringbauer/git/hapROH/Notebooks/Vignettes\n"
     ]
    }
   ],
   "source": [
    "### Fill in your own path here!\n",
    "path = \"/mnt/archgen/users/hringbauer/git/hapROH/Notebooks/Vignettes/\"  # The Path to Package Midway Cluster\n",
    "\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory. Should be HAPSBURG/Notebooks/ParallelRuns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1) Download the data we will use as an example here\n",
    "If you don't have it already, you will need to get the target and reference data. Here is how:\n",
    "\n",
    "### Get Target Data\n",
    "We will download an example eigenstrat available via the Reich lab homepage. \n",
    "It contains a handful of individuals from Chalcolithic Levant (zipped, its's ~20mb then).\n",
    "You can download it from:\n",
    "`https://reich.hms.harvard.edu/sites/reich.hms.harvard.edu/files/inline-files/Levant_ChL.tar.gz`\n",
    "Download and unpack in the folder you want it (unpack to .geno, .ind and .snp file).\n",
    "\n",
    "### Get Reference Data\n",
    "You can download the reference data from \n",
    "`https://www.dropbox.com/s/0qhjgo1npeih0bw/1000g1240khdf5.tar.gz?dl=0`\n",
    "\n",
    "This data is key for running hapROH - it is the 1000 Genome data downsampled to bi-allelic 1240k SNPs, from 5008 global reference haplotypes (Attention: file is ca. 800 mb big). \n",
    "\n",
    "There is also a metafile datatable that contains information about the reference panel. This table is necessary to run hapROH - do not forget to copy that over too in addition to data for every chromosome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2) Call ROH\n",
    "Now we will run the core part of the package, a wrapper function for the \n",
    "core ROH calling machinery - `hapsb_ind`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test calling ROH on single chromosome from Individual with full output printed\n",
    "Set the location of the reference hdf5 and the target eigenstrat path correctly!\n",
    "\n",
    "Input: Reference haplotype file (hdf5, containing genetic map), reference metafile, pseudo-haploid Eigenstrat\n",
    "Output: Output file in folder `folder_out`, importantly a \n",
    "\n",
    "See example below, you need to set `path_targets` and `h5_path1000G`\n",
    "\n",
    "logfile=False -> All output is printed into sys.out  \n",
    "combine=False -> No individual .csv is created from chromosome .csvs "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Minimal Version\n",
    "A version where the complex parameters a normal user does not have to set hidden (i.e. set to default values that work well for 1240K data).\n",
    "hapROH is tested for a wide variety of human aDNA data, and the default parameters likely have you covered if you run hapROH on <50 ky old 1240K data.\n",
    "\n",
    "This function will call ROH of indivdiual `iid` in the target eigenstrat given in `path_targets`. We also set the path to the reference hdf5 file (`h5_path1000G`) and the metafile of the reference (`meta_path_ref`), and, importantly, into which folder to save the output to (`folder_out`).\n",
    "\n",
    "Let us run the date for chromosome 20 (range(20,21) selects a list of length 1: [20,]), and using 1 processor (`processes=1`) as well as printing all text output (`output=True`) into the connsole (`logfile=False`). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I1178\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr20.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 22 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29078\n",
      "Nr of Matching Refs: 29078 / 29078\n",
      "Ref/Alt Matching: 29053 / 29078\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29053 / 29078\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 19815 / 29053\n",
      "Fraction SNPs covered: 0.6820\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./ExampleOutput/I1178/chr20/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./ExampleOutput/I1178/chr20/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI1178\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "downsample\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./ExampleData/Levant_ChL\n",
      "h5_path1000g\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\n",
      "folder_out\t./ExampleOutput/\n",
      "prefix_out\t\n",
      "c\t0.0\n",
      "conPop\t['CEU']\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0.005\n",
      "roh_min_l_initial\t0.02\n",
      "roh_min_l_final\t0.04\n",
      "min_len1\t0.02\n",
      "min_len2\t0.04\n",
      "cutoff_post\t0.999\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 159\n",
      "Maximum Gap: 0.4744 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 19815)\n",
      "Loaded Observations:\n",
      "(2, 19815)\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1154.461696 mB\n",
      "Total Log likelihood: -5803.614\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./ExampleOutput/I1178/chr20/.\n",
      "Successfully loaded for PP. from ./ExampleOutput/I1178/chr20/\n",
      "Fraction Markers in ROH: 0.4818\n",
      "Called n=1 ROH Blocks > 4.0 cM\n",
      "Longest Block: 26.10 cM\n",
      "Successfully saved to ./ExampleOutput/I1178/chr20/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "hapsb_ind(iid=\"I1178\", chs=range(20, 21), \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "          meta_path_ref=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\", \n",
    "          folder_out='./ExampleOutput/',  # Folder where you want to save the results to \n",
    "          processes=1, output=True,\n",
    "          readcounts=False, logfile=False, combine=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Power version [does same as above, but shows more parameters that could be changed]\n",
    "Shows most of the parameters that you can set. Warning: Try to understand what you are doing.\n",
    "E.g. the transition parameters are optimized for 1240k data already - changing them can have ...fun consequences.\n",
    "\n",
    "The parameters are explained in the doc string of the function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I1178\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr20.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 22 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29078\n",
      "Nr of Matching Refs: 29078 / 29078\n",
      "Ref/Alt Matching: 29053 / 29078\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29053 / 29078\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 19815 / 29053\n",
      "Fraction SNPs covered: 0.6820\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./ExampleOutput/I1178/chr20/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./ExampleOutput/I1178/chr20/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI1178\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "downsample\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./ExampleData/Levant_ChL\n",
      "h5_path1000g\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\n",
      "folder_out\t./ExampleOutput/\n",
      "prefix_out\t\n",
      "c\t0.0\n",
      "conPop\t['CEU']\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0.005\n",
      "roh_min_l_initial\t0.02\n",
      "roh_min_l_final\t0.05\n",
      "min_len1\t0.02\n",
      "min_len2\t0.04\n",
      "cutoff_post\t0.999\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 159\n",
      "Maximum Gap: 0.4744 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 19815)\n",
      "Loaded Observations:\n",
      "(2, 19815)\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1141.960704 mB\n",
      "Total Log likelihood: -5803.614\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./ExampleOutput/I1178/chr20/.\n",
      "Successfully loaded for PP. from ./ExampleOutput/I1178/chr20/\n",
      "Fraction Markers in ROH: 0.4818\n",
      "Called n=1 ROH Blocks > 5.0 cM\n",
      "Longest Block: 26.10 cM\n",
      "Successfully saved to ./ExampleOutput/I1178/chr20/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "hapsb_ind(iid=\"I1178\", chs=range(20, 21), \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "          meta_path_ref=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\", \n",
    "          folder_out='./ExampleOutput/', prefix_out='', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, save_fp=False, \n",
    "          n_ref=2504, diploid_ref=True, exclude_pops=[], readcounts=False, random_allele=True,\n",
    "          c=0.0, conPop=[\"CEU\"], roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.00, \n",
    "          cutoff_post = 0.999, max_gap=0.005, roh_min_l_initial = 0.02, roh_min_l_final = 0.05,\n",
    "          min_len1 = 0.02, min_len2 = 0.04, logfile=False, combine=False, file_result=\"_roh_full.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example run of whole individual (all chromosomes) with output to logfile\n",
    "This example runs a whole individual in parallel, with the output send to a logfile\n",
    "\n",
    "Attention: hapROH uses quite a bit of RAM, for an Eigenstrat it can be spiking up to 6gb for a long chromosome with most SNPs covered - allocate memory accoringly\n",
    "when running multiple processes, or set that number lower!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual I1178...\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr2/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr4/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr6/hmm_run_log.txt\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>StartM</th>\n",
       "      <th>EndM</th>\n",
       "      <th>length</th>\n",
       "      <th>lengthM</th>\n",
       "      <th>iid</th>\n",
       "      <th>ch</th>\n",
       "      <th>StartPosGRCh37</th>\n",
       "      <th>EndPosGRCh37</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>939</td>\n",
       "      <td>3857</td>\n",
       "      <td>0.103408</td>\n",
       "      <td>0.286113</td>\n",
       "      <td>2918</td>\n",
       "      <td>0.182705</td>\n",
       "      <td>I1178</td>\n",
       "      <td>1</td>\n",
       "      <td>4367323</td>\n",
       "      <td>13864528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39593</td>\n",
       "      <td>45713</td>\n",
       "      <td>1.989456</td>\n",
       "      <td>2.278305</td>\n",
       "      <td>6120</td>\n",
       "      <td>0.288849</td>\n",
       "      <td>I1178</td>\n",
       "      <td>1</td>\n",
       "      <td>183297554</td>\n",
       "      <td>211079618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38445</td>\n",
       "      <td>40912</td>\n",
       "      <td>1.766828</td>\n",
       "      <td>1.905809</td>\n",
       "      <td>2467</td>\n",
       "      <td>0.138981</td>\n",
       "      <td>I1178</td>\n",
       "      <td>2</td>\n",
       "      <td>163623169</td>\n",
       "      <td>175599106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3216</td>\n",
       "      <td>9660</td>\n",
       "      <td>0.236136</td>\n",
       "      <td>0.555893</td>\n",
       "      <td>6444</td>\n",
       "      <td>0.319757</td>\n",
       "      <td>I1178</td>\n",
       "      <td>3</td>\n",
       "      <td>9248379</td>\n",
       "      <td>32219726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19509</td>\n",
       "      <td>40280</td>\n",
       "      <td>0.917425</td>\n",
       "      <td>1.828643</td>\n",
       "      <td>20771</td>\n",
       "      <td>0.911218</td>\n",
       "      <td>I1178</td>\n",
       "      <td>3</td>\n",
       "      <td>68600488</td>\n",
       "      <td>174410136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41385</td>\n",
       "      <td>42691</td>\n",
       "      <td>1.903147</td>\n",
       "      <td>1.993923</td>\n",
       "      <td>1306</td>\n",
       "      <td>0.090776</td>\n",
       "      <td>I1178</td>\n",
       "      <td>3</td>\n",
       "      <td>180638648</td>\n",
       "      <td>186587417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1758</td>\n",
       "      <td>4365</td>\n",
       "      <td>0.127779</td>\n",
       "      <td>0.295442</td>\n",
       "      <td>2607</td>\n",
       "      <td>0.167663</td>\n",
       "      <td>I1178</td>\n",
       "      <td>4</td>\n",
       "      <td>5995255</td>\n",
       "      <td>13723484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24314</td>\n",
       "      <td>26821</td>\n",
       "      <td>1.306104</td>\n",
       "      <td>1.444817</td>\n",
       "      <td>2507</td>\n",
       "      <td>0.138713</td>\n",
       "      <td>I1178</td>\n",
       "      <td>4</td>\n",
       "      <td>122673274</td>\n",
       "      <td>139861418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>3592</td>\n",
       "      <td>0.000381</td>\n",
       "      <td>0.217287</td>\n",
       "      <td>3592</td>\n",
       "      <td>0.216906</td>\n",
       "      <td>I1178</td>\n",
       "      <td>5</td>\n",
       "      <td>26366</td>\n",
       "      <td>9607318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>7940</td>\n",
       "      <td>0.003530</td>\n",
       "      <td>0.443168</td>\n",
       "      <td>7940</td>\n",
       "      <td>0.439638</td>\n",
       "      <td>I1178</td>\n",
       "      <td>6</td>\n",
       "      <td>203909</td>\n",
       "      <td>23616661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>32635</td>\n",
       "      <td>44713</td>\n",
       "      <td>1.294776</td>\n",
       "      <td>1.920323</td>\n",
       "      <td>12078</td>\n",
       "      <td>0.625547</td>\n",
       "      <td>I1178</td>\n",
       "      <td>6</td>\n",
       "      <td>128207306</td>\n",
       "      <td>170906415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>399</td>\n",
       "      <td>5239</td>\n",
       "      <td>0.035357</td>\n",
       "      <td>0.384504</td>\n",
       "      <td>4840</td>\n",
       "      <td>0.349147</td>\n",
       "      <td>I1178</td>\n",
       "      <td>7</td>\n",
       "      <td>2298227</td>\n",
       "      <td>21852852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10565</td>\n",
       "      <td>15042</td>\n",
       "      <td>0.618036</td>\n",
       "      <td>0.790791</td>\n",
       "      <td>4477</td>\n",
       "      <td>0.172755</td>\n",
       "      <td>I1178</td>\n",
       "      <td>7</td>\n",
       "      <td>38536202</td>\n",
       "      <td>55161156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14654</td>\n",
       "      <td>21043</td>\n",
       "      <td>0.856262</td>\n",
       "      <td>1.104058</td>\n",
       "      <td>6389</td>\n",
       "      <td>0.247796</td>\n",
       "      <td>I1178</td>\n",
       "      <td>9</td>\n",
       "      <td>85960840</td>\n",
       "      <td>109244902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18956</td>\n",
       "      <td>21259</td>\n",
       "      <td>0.987105</td>\n",
       "      <td>1.075941</td>\n",
       "      <td>2303</td>\n",
       "      <td>0.088836</td>\n",
       "      <td>I1178</td>\n",
       "      <td>10</td>\n",
       "      <td>74885920</td>\n",
       "      <td>82647730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9584</td>\n",
       "      <td>11247</td>\n",
       "      <td>0.567689</td>\n",
       "      <td>0.625520</td>\n",
       "      <td>1663</td>\n",
       "      <td>0.057831</td>\n",
       "      <td>I1178</td>\n",
       "      <td>12</td>\n",
       "      <td>41695580</td>\n",
       "      <td>48935999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9775</td>\n",
       "      <td>11267</td>\n",
       "      <td>0.587189</td>\n",
       "      <td>0.668740</td>\n",
       "      <td>1492</td>\n",
       "      <td>0.081551</td>\n",
       "      <td>I1178</td>\n",
       "      <td>13</td>\n",
       "      <td>61328047</td>\n",
       "      <td>72428647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2863</td>\n",
       "      <td>4101</td>\n",
       "      <td>0.432307</td>\n",
       "      <td>0.523272</td>\n",
       "      <td>1238</td>\n",
       "      <td>0.090965</td>\n",
       "      <td>I1178</td>\n",
       "      <td>15</td>\n",
       "      <td>33936879</td>\n",
       "      <td>39111829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4130</td>\n",
       "      <td>11683</td>\n",
       "      <td>0.529911</td>\n",
       "      <td>0.837226</td>\n",
       "      <td>7553</td>\n",
       "      <td>0.307315</td>\n",
       "      <td>I1178</td>\n",
       "      <td>15</td>\n",
       "      <td>39181802</td>\n",
       "      <td>68677349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1501</td>\n",
       "      <td>7649</td>\n",
       "      <td>0.114700</td>\n",
       "      <td>0.481877</td>\n",
       "      <td>6148</td>\n",
       "      <td>0.367177</td>\n",
       "      <td>I1178</td>\n",
       "      <td>16</td>\n",
       "      <td>5664127</td>\n",
       "      <td>24534255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20076</td>\n",
       "      <td>23380</td>\n",
       "      <td>1.117170</td>\n",
       "      <td>1.340230</td>\n",
       "      <td>3304</td>\n",
       "      <td>0.223060</td>\n",
       "      <td>I1178</td>\n",
       "      <td>16</td>\n",
       "      <td>83103233</td>\n",
       "      <td>90141477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>825</td>\n",
       "      <td>4488</td>\n",
       "      <td>0.082527</td>\n",
       "      <td>0.337067</td>\n",
       "      <td>3663</td>\n",
       "      <td>0.254540</td>\n",
       "      <td>I1178</td>\n",
       "      <td>17</td>\n",
       "      <td>2906710</td>\n",
       "      <td>13152867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18316</td>\n",
       "      <td>20193</td>\n",
       "      <td>1.061995</td>\n",
       "      <td>1.177033</td>\n",
       "      <td>1877</td>\n",
       "      <td>0.115038</td>\n",
       "      <td>I1178</td>\n",
       "      <td>18</td>\n",
       "      <td>73099261</td>\n",
       "      <td>78002055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>315</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.056503</td>\n",
       "      <td>315</td>\n",
       "      <td>0.056465</td>\n",
       "      <td>I1178</td>\n",
       "      <td>19</td>\n",
       "      <td>267614</td>\n",
       "      <td>1550650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>341</td>\n",
       "      <td>3941</td>\n",
       "      <td>0.063010</td>\n",
       "      <td>0.397439</td>\n",
       "      <td>3600</td>\n",
       "      <td>0.334429</td>\n",
       "      <td>I1178</td>\n",
       "      <td>19</td>\n",
       "      <td>1735218</td>\n",
       "      <td>16172739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7886</td>\n",
       "      <td>11980</td>\n",
       "      <td>0.615688</td>\n",
       "      <td>0.920067</td>\n",
       "      <td>4094</td>\n",
       "      <td>0.304379</td>\n",
       "      <td>I1178</td>\n",
       "      <td>19</td>\n",
       "      <td>36123403</td>\n",
       "      <td>53809915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4270</td>\n",
       "      <td>10948</td>\n",
       "      <td>0.316092</td>\n",
       "      <td>0.577081</td>\n",
       "      <td>6678</td>\n",
       "      <td>0.260989</td>\n",
       "      <td>I1178</td>\n",
       "      <td>20</td>\n",
       "      <td>12276356</td>\n",
       "      <td>39406575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>585</td>\n",
       "      <td>0.018961</td>\n",
       "      <td>0.072243</td>\n",
       "      <td>574</td>\n",
       "      <td>0.053282</td>\n",
       "      <td>I1178</td>\n",
       "      <td>21</td>\n",
       "      <td>15454659</td>\n",
       "      <td>17832871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>592</td>\n",
       "      <td>958</td>\n",
       "      <td>0.085863</td>\n",
       "      <td>0.126019</td>\n",
       "      <td>366</td>\n",
       "      <td>0.040156</td>\n",
       "      <td>I1178</td>\n",
       "      <td>21</td>\n",
       "      <td>17885987</td>\n",
       "      <td>19557045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10411</td>\n",
       "      <td>11924</td>\n",
       "      <td>0.634021</td>\n",
       "      <td>0.740478</td>\n",
       "      <td>1513</td>\n",
       "      <td>0.106457</td>\n",
       "      <td>I1178</td>\n",
       "      <td>22</td>\n",
       "      <td>48008036</td>\n",
       "      <td>51165664</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Start    End    StartM      EndM length   lengthM    iid  ch  \\\n",
       "0    939   3857  0.103408  0.286113   2918  0.182705  I1178   1   \n",
       "1  39593  45713  1.989456  2.278305   6120  0.288849  I1178   1   \n",
       "0  38445  40912  1.766828  1.905809   2467  0.138981  I1178   2   \n",
       "0   3216   9660  0.236136  0.555893   6444  0.319757  I1178   3   \n",
       "1  19509  40280  0.917425  1.828643  20771  0.911218  I1178   3   \n",
       "2  41385  42691  1.903147  1.993923   1306  0.090776  I1178   3   \n",
       "0   1758   4365  0.127779  0.295442   2607  0.167663  I1178   4   \n",
       "1  24314  26821  1.306104  1.444817   2507  0.138713  I1178   4   \n",
       "0      0   3592  0.000381  0.217287   3592  0.216906  I1178   5   \n",
       "0      0   7940  0.003530  0.443168   7940  0.439638  I1178   6   \n",
       "1  32635  44713  1.294776  1.920323  12078  0.625547  I1178   6   \n",
       "0    399   5239  0.035357  0.384504   4840  0.349147  I1178   7   \n",
       "1  10565  15042  0.618036  0.790791   4477  0.172755  I1178   7   \n",
       "0  14654  21043  0.856262  1.104058   6389  0.247796  I1178   9   \n",
       "0  18956  21259  0.987105  1.075941   2303  0.088836  I1178  10   \n",
       "0   9584  11247  0.567689  0.625520   1663  0.057831  I1178  12   \n",
       "0   9775  11267  0.587189  0.668740   1492  0.081551  I1178  13   \n",
       "0   2863   4101  0.432307  0.523272   1238  0.090965  I1178  15   \n",
       "1   4130  11683  0.529911  0.837226   7553  0.307315  I1178  15   \n",
       "0   1501   7649  0.114700  0.481877   6148  0.367177  I1178  16   \n",
       "1  20076  23380  1.117170  1.340230   3304  0.223060  I1178  16   \n",
       "0    825   4488  0.082527  0.337067   3663  0.254540  I1178  17   \n",
       "0  18316  20193  1.061995  1.177033   1877  0.115038  I1178  18   \n",
       "0      0    315  0.000038  0.056503    315  0.056465  I1178  19   \n",
       "1    341   3941  0.063010  0.397439   3600  0.334429  I1178  19   \n",
       "2   7886  11980  0.615688  0.920067   4094  0.304379  I1178  19   \n",
       "0   4270  10948  0.316092  0.577081   6678  0.260989  I1178  20   \n",
       "0     11    585  0.018961  0.072243    574  0.053282  I1178  21   \n",
       "1    592    958  0.085863  0.126019    366  0.040156  I1178  21   \n",
       "0  10411  11924  0.634021  0.740478   1513  0.106457  I1178  22   \n",
       "\n",
       "  StartPosGRCh37 EndPosGRCh37  \n",
       "0        4367323     13864528  \n",
       "1      183297554    211079618  \n",
       "0      163623169    175599106  \n",
       "0        9248379     32219726  \n",
       "1       68600488    174410136  \n",
       "2      180638648    186587417  \n",
       "0        5995255     13723484  \n",
       "1      122673274    139861418  \n",
       "0          26366      9607318  \n",
       "0         203909     23616661  \n",
       "1      128207306    170906415  \n",
       "0        2298227     21852852  \n",
       "1       38536202     55161156  \n",
       "0       85960840    109244902  \n",
       "0       74885920     82647730  \n",
       "0       41695580     48935999  \n",
       "0       61328047     72428647  \n",
       "0       33936879     39111829  \n",
       "1       39181802     68677349  \n",
       "0        5664127     24534255  \n",
       "1       83103233     90141477  \n",
       "0        2906710     13152867  \n",
       "0       73099261     78002055  \n",
       "0         267614      1550650  \n",
       "1        1735218     16172739  \n",
       "2       36123403     53809915  \n",
       "0       12276356     39406575  \n",
       "0       15454659     17832871  \n",
       "1       17885987     19557045  \n",
       "0       48008036     51165664  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapsb_ind(iid=\"I1178\", chs=range(1,23), processes=6, \n",
    "          path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "          meta_path_ref=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\", \n",
    "          folder_out='./ExampleOutput/', prefix_out='', \n",
    "          e_model=\"haploid\", p_model=\"Eigenstrat\",\n",
    "          random_allele=True, readcounts=False,\n",
    "          delete=False, logfile=True, combine=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run multiple Individuals\n",
    "Here as a loop over individuals. In practice I run it as parallelized sbatch job (setting processes to 1 there) - this way you can parallelize it with 1 processor per indivual. That is a useful level of parallelization, as the individual output files will be combined in seperate step below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing Individual: I1178\n",
      "Doing Individual I1178...\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr1/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr6/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1178/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1178/chr5/hmm_run_log.txt\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I0644\n",
      "Doing Individual I0644...\n",
      "Set Output Log path: ./ExampleOutput/I0644/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr4/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I0644/chr5/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I0644/chr6/hmm_run_log.txt\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1160\n",
      "Doing Individual I1160...\n",
      "Set Output Log path: ./ExampleOutput/I1160/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1160/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1160/chr3/hmm_run_log.txt\n",
      "\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1160/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1160/chr6/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1152\n",
      "Doing Individual I1152...\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr2/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1152/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr3/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr6/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1152/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1168\n",
      "Doing Individual I1168...\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr6/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr2/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1168/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1166\n",
      "Doing Individual I1166...\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr2/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr6/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1166/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1170\n",
      "Doing Individual I1170...\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr2/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr6/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1170/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1165\n",
      "Doing Individual I1165...\n",
      "Set Output Log path: ./ExampleOutput/I1165/chr4/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr3/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1165/chr1/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr2/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1165/chr6/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1165/chr5/hmm_run_log.txt\n",
      "\n",
      "Combining Information for 22 Chromosomes...\n",
      "Doing Individual: I1182\n",
      "Doing Individual I1182...\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr2/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr4/hmm_run_log.txt\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr3/hmm_run_log.txtSet Output Log path: ./ExampleOutput/I1182/chr6/hmm_run_log.txt\n",
      "\n",
      "Set Output Log path: ./ExampleOutput/I1182/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n"
     ]
    }
   ],
   "source": [
    "### This are all individuals with 400k SNPs covered\n",
    "iids = ['I1178', 'I0644', 'I1160', 'I1152', 'I1168', 'I1166', 'I1170', 'I1165', 'I1182']\n",
    "\n",
    "for iid in iids:\n",
    "    print(f\"Doing Individual: {iid}\")\n",
    "    hapsb_ind(iid=iid, chs=range(1,23), processes=6, \n",
    "              path_targets='./ExampleData/Levant_ChL', # The path before the .ind, .snp, .geno\n",
    "              h5_path1000g=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "              meta_path_ref=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\", \n",
    "              folder_out='./ExampleOutput/', prefix_out='', \n",
    "              e_model=\"haploid\", p_model=\"EigenstratPacked\", n_ref=2504,\n",
    "              random_allele=True, readcounts=False,\n",
    "              delete=False, logfile=True, combine=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Postprocess Results into one results.csv (copying in Meta Data)\n",
    "Take indivdiual output .csvs and combine into one big results .csv\n",
    "Merging of output gaps, and ROH>x cM happens here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hapsburg.PackagesSupport.pp_individual_roh_csvs import pp_individual_roh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create Example Meta File\n",
    "This file will create a minimal meta data file (some plotting functions use individual meta information, such as age),\n",
    "which is then merged into the results when combining individuals.\n",
    "\n",
    "If you have metadata available for your dataset, you can prepare a comma seperated csv table with these headers!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "iids = ['I1178', 'I0644', 'I1160', \n",
    "        'I1152', 'I1168', 'I1166', \n",
    "        'I1170', 'I1165', 'I1182']\n",
    "df = pd.DataFrame({\"iid\":iids})\n",
    "\n",
    "df[\"age\"] = 5950\n",
    "df[\"clst\"] = \"Israel_C\"\n",
    "df[\"lat\"] = 32.974167\n",
    "df[\"lon\"] = 35.331389\n",
    "\n",
    "df.to_csv(\"./ExampleData/meta_blank.csv\", \n",
    "          sep=\",\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine individual output files into a single ROH output File\n",
    "Combines meta file, individual ROH output files\n",
    "into a summary table, contining statistical information about ROH for each indiviual as a row.\n",
    "\n",
    "This is the table that a lot of the downstream plotting software uses.\n",
    "\n",
    "Creating this table also does additional post-processing of the ROH, in particular merging gaps between ROH. Every gap of length gap [in cM] between two ROH of >min_len1 and >min_len2 is merged.\n",
    "\n",
    "This can be already done in hapsb_ind, but can be also done here. Doing it only in the end is useful e.g. in cases where one wants to try different post-processing and gap-merging:\n",
    "One could set the gap to 0 in hapsb_ind and then here try different values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 9 / 9 Individuals from Meta\n",
      "Saved to: ./ExampleOutput/combined_roh05.csv\n",
      "CPU times: user 294 ms, sys: 14.7 ms, total: 309 ms\n",
      "Wall time: 336 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "### Postprocess the two Individuals from above and combine into one results .csv\n",
    "iids = ['I1178', 'I0644', 'I1160', 'I1152', \n",
    "        'I1168', 'I1166', 'I1170', 'I1165', 'I1182']\n",
    "\n",
    "df1 = pp_individual_roh(iids, meta_path=\"./ExampleData/meta_blank.csv\", \n",
    "                        base_folder=\"./ExampleOutput/\",\n",
    "                        save_path=\"./ExampleOutput/combined_roh05.csv\", \n",
    "                        output=False, min_cm=[4, 8, 12, 20], snp_cm=50, \n",
    "                        gap=0.5, min_len1=2.0, min_len2=4.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the output table\n",
    "As you can see, one indivdiual has extremeley high amounts of ROH. More about interpreting and visualizing that in `./plotROH_vignette.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>max_roh</th>\n",
       "      <th>iid</th>\n",
       "      <th>pop</th>\n",
       "      <th>sum_roh&gt;4</th>\n",
       "      <th>n_roh&gt;4</th>\n",
       "      <th>sum_roh&gt;8</th>\n",
       "      <th>n_roh&gt;8</th>\n",
       "      <th>sum_roh&gt;12</th>\n",
       "      <th>n_roh&gt;12</th>\n",
       "      <th>sum_roh&gt;20</th>\n",
       "      <th>n_roh&gt;20</th>\n",
       "      <th>age</th>\n",
       "      <th>clst</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>91.121798</td>\n",
       "      <td>I1178</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>703.292088</td>\n",
       "      <td>30</td>\n",
       "      <td>682.518689</td>\n",
       "      <td>26</td>\n",
       "      <td>625.156411</td>\n",
       "      <td>20</td>\n",
       "      <td>545.074702</td>\n",
       "      <td>15</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9.192801</td>\n",
       "      <td>I1170</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>26.837993</td>\n",
       "      <td>4</td>\n",
       "      <td>18.287802</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.879100</td>\n",
       "      <td>I1166</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>9.593799</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.202509</td>\n",
       "      <td>I1160</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>8.283711</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I0644</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1152</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1168</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1165</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>I1182</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>5950</td>\n",
       "      <td>Israel_C</td>\n",
       "      <td>32.974167</td>\n",
       "      <td>35.331389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     max_roh    iid       pop   sum_roh>4  n_roh>4   sum_roh>8  n_roh>8  \\\n",
       "0  91.121798  I1178  Israel_C  703.292088       30  682.518689       26   \n",
       "1   9.192801  I1170  Israel_C   26.837993        4   18.287802        2   \n",
       "2   4.879100  I1166  Israel_C    9.593799        2    0.000000        0   \n",
       "3   4.202509  I1160  Israel_C    8.283711        2    0.000000        0   \n",
       "4   0.000000  I0644  Israel_C    0.000000        0    0.000000        0   \n",
       "5   0.000000  I1152  Israel_C    0.000000        0    0.000000        0   \n",
       "6   0.000000  I1168  Israel_C    0.000000        0    0.000000        0   \n",
       "7   0.000000  I1165  Israel_C    0.000000        0    0.000000        0   \n",
       "8   0.000000  I1182  Israel_C    0.000000        0    0.000000        0   \n",
       "\n",
       "   sum_roh>12  n_roh>12  sum_roh>20  n_roh>20   age      clst        lat  \\\n",
       "0  625.156411        20  545.074702        15  5950  Israel_C  32.974167   \n",
       "1    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "2    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "3    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "4    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "5    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "6    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "7    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "8    0.000000         0    0.000000         0  5950  Israel_C  32.974167   \n",
       "\n",
       "         lon  \n",
       "0  35.331389  \n",
       "1  35.331389  \n",
       "2  35.331389  \n",
       "3  35.331389  \n",
       "4  35.331389  \n",
       "5  35.331389  \n",
       "6  35.331389  \n",
       "7  35.331389  \n",
       "8  35.331389  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
