{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Estimating Contamination from ROH Blocks\n",
    "In this notebook we walk you through how to use hapCON_auto.py to jointly estimate ROH and contamination rates. The algorithm is described in detail in xxx. For a short introduction and application range of this method, please refer to our [readthedocs site](https://haproh.readthedocs.io/en/latest/hapROH_with_contamination.html).\n",
    "\n",
    "To run this notebook, you need install the hapROH package (version xxx or higher), [BamTable](https://bioinf.eva.mpg.de/BamTable/) and have samtools in your path. In principle, one only need one of BamTable/samtools, as our method can take either as input. However, we recommend using BamTable in practice as it provides more flexibility when preprocessing your BAM files. \n",
    "\n",
    "We use a test BAM file of 0.25x coverage as an example in this tutorial. It was a simulated contaminated BAM files using I4432 (a Latvia hunter-gatherer, 6072-5920 calBCE, whole genome sequenced to average depth 21.23x) as the endogenous source and I5319 (Alaska, USA, 1050-1400 calCE, whole genome sequenced to average depth 21.34x) as the contamination source. The original data was obtained from [Allen Ancient Genome Diversity Project](https://reich.hms.harvard.edu/ancient-genome-diversity-project). You can download our test BAM file from xxx. This endogenous source (I4432) has about ~50cM of ROH and has 10% simulated contamination.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's first run ROH contamination estimates using [BamTable](https://bioinf.eva.mpg.de/BamTable/) as input. <span style='color:red'>Please change the path to BamTable binary and BED files according to your own setup. We have assumed that the test BAM file is in directory \"./Data\"</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# please change the path to BamTable and BED file according to your own setup!\n",
    "!mkdir ./Data/BamTable\n",
    "!for ch in {1..22}; do /mnt/archgen/users/yilei/bin/BamTable-2.0.3/BamTable -f /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/1240kChr$ch.bed -b -A ./Data/ROH_contam.test.bam > ./Data/BamTable/test.chr$ch.BamTable; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style='color:red'>Please make sure that the BamTable output files are named as \"$iid.chr{1..22}.BamTable\"</span> The sample iid is specified via the command line argument -i. Now we are ready to run hapCON_auto.py. It is not very fast (primarily because it has to load the reference panels multiple times) and will take about 12min. The primary output will be named as $iid.$prefix.hapCON_ROH.txt. It contains the final contamination estimates and several other useful information, for example, how many ROH blocks are used in estimation and their total length. At the end of the run, we make a final ROH calling using the estimated contamination rate. The result of this final ROH calling is stored in directory ./hapRoh_iter. You can use this as your sample's final ROH result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving hdf5 files in ./Data/hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr7.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr5.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr3.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr8.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr1.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr6.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr4.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr9.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr13.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr11.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr14.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr10.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr15.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr17.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr12.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr2.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr18.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr16.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr21.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr19.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr22.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr20.hdf5\n",
      "finished reading input files, takes 29.483s\n",
      "estimated genotyping error: 0.001982544988581601\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.510303\t0.635444\n",
      "chr6\t1.042793\t1.131743\n",
      "chr9\t0.96028\t1.029503\n",
      "a total of 3 ROH blocks passing filtering threshold found, total length after trimming: 25.331cM.\n",
      "3 hmm models loaded, takes 79.404s\n",
      "hapsb_multiChunk_preload takes 2.4557347297668457\n",
      "hapsb_multiChunk_preload takes 2.6880581378936768\n",
      "hapsb_multiChunk_preload takes 2.6434667110443115\n",
      "hapsb_multiChunk_preload takes 2.6561520099639893\n",
      "hapsb_multiChunk_preload takes 2.7429139614105225\n",
      "hapsb_multiChunk_preload takes 2.6487369537353516\n",
      "hapsb_multiChunk_preload takes 2.529193162918091\n",
      "hapsb_multiChunk_preload takes 2.6594057083129883\n",
      "hapsb_multiChunk_preload takes 2.492086410522461\n",
      "hapsb_multiChunk_preload takes 2.5559377670288086\n",
      "hapsb_multiChunk_preload takes 2.584928035736084\n",
      "hapsb_multiChunk_preload takes 2.5916426181793213\n",
      "hapsb_multiChunk_preload takes 2.6633493900299072\n",
      "hapsb_multiChunk_preload takes 2.6252806186676025\n",
      "hapsb_multiChunk_preload takes 2.641185760498047\n",
      "hapsb_multiChunk_preload takes 2.5257718563079834\n",
      "hapsb_multiChunk_preload takes 2.561204195022583\n",
      "hapsb_multiChunk_preload takes 2.5037472248077393\n",
      "hapsb_multiChunk_preload takes 2.700666666030884\n",
      "hapsb_multiChunk_preload takes 2.6468560695648193\n",
      "hapsb_multiChunk_preload takes 2.533064126968384\n",
      "hapsb_multiChunk_preload takes 2.5486466884613037\n",
      "hapsb_multiChunk_preload takes 2.640960693359375\n",
      "estimated contamination rate 0.09847 greater than 0.05, start iteration.\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.472396\t0.639327\n",
      "chr2\t2.220766\t2.279183\n",
      "chr6\t1.042141\t1.195667\n",
      "chr9\t0.96028\t1.054782\n",
      "a total of 4 ROH blocks passing filtering threshold found, total length after trimming: 43.338cM.\n",
      "4 hmm models loaded, takes 85.439s\n",
      "hapsb_multiChunk_preload takes 3.282395124435425\n",
      "hapsb_multiChunk_preload takes 3.1412711143493652\n",
      "hapsb_multiChunk_preload takes 3.3085618019104004\n",
      "hapsb_multiChunk_preload takes 3.3128082752227783\n",
      "hapsb_multiChunk_preload takes 3.096367835998535\n",
      "hapsb_multiChunk_preload takes 3.2934508323669434\n",
      "hapsb_multiChunk_preload takes 3.3393189907073975\n",
      "hapsb_multiChunk_preload takes 3.2503585815429688\n",
      "hapsb_multiChunk_preload takes 3.1598961353302\n",
      "hapsb_multiChunk_preload takes 3.1578924655914307\n",
      "hapsb_multiChunk_preload takes 3.099175214767456\n",
      "hapsb_multiChunk_preload takes 3.0210368633270264\n",
      "hapsb_multiChunk_preload takes 3.179079294204712\n",
      "hapsb_multiChunk_preload takes 3.2316830158233643\n",
      "hapsb_multiChunk_preload takes 3.305864095687866\n",
      "hapsb_multiChunk_preload takes 3.0722219944000244\n",
      "hapsb_multiChunk_preload takes 3.2693564891815186\n",
      "hapsb_multiChunk_preload takes 3.29129695892334\n",
      "hapsb_multiChunk_preload takes 3.3282132148742676\n",
      "hapsb_multiChunk_preload takes 3.156092882156372\n",
      "hapsb_multiChunk_preload takes 3.118720531463623\n",
      "iteration 2 done, prev contam: 0.09847, current contam: 0.120581\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.472343\t0.644013\n",
      "chr2\t2.220766\t2.279213\n",
      "chr6\t1.039623\t1.196152\n",
      "chr9\t0.958626\t1.056485\n",
      "chr19\t0.076272\t0.166039\n",
      "a total of 5 ROH blocks passing filtering threshold found, total length after trimming: 52.427cM.\n",
      "5 hmm models loaded, takes 82.533s\n",
      "hapsb_multiChunk_preload takes 3.236881971359253\n",
      "hapsb_multiChunk_preload takes 3.2390639781951904\n",
      "hapsb_multiChunk_preload takes 3.18778920173645\n",
      "hapsb_multiChunk_preload takes 3.2789247035980225\n",
      "hapsb_multiChunk_preload takes 3.1971018314361572\n",
      "hapsb_multiChunk_preload takes 3.2682995796203613\n",
      "hapsb_multiChunk_preload takes 3.23363995552063\n",
      "hapsb_multiChunk_preload takes 3.289886951446533\n",
      "hapsb_multiChunk_preload takes 3.392472743988037\n",
      "hapsb_multiChunk_preload takes 3.1809494495391846\n",
      "hapsb_multiChunk_preload takes 3.2952799797058105\n",
      "hapsb_multiChunk_preload takes 3.514385223388672\n",
      "hapsb_multiChunk_preload takes 3.460145950317383\n",
      "hapsb_multiChunk_preload takes 3.2590172290802\n",
      "hapsb_multiChunk_preload takes 3.3799638748168945\n",
      "hapsb_multiChunk_preload takes 3.449328660964966\n",
      "hapsb_multiChunk_preload takes 3.3748834133148193\n",
      "hapsb_multiChunk_preload takes 3.196953296661377\n",
      "hapsb_multiChunk_preload takes 3.6498992443084717\n",
      "hapsb_multiChunk_preload takes 3.318481922149658\n",
      "hapsb_multiChunk_preload takes 3.259401798248291\n",
      "iteration 3 done, prev contam: 0.120581, current contam: 0.121682\n",
      "contamination rate converged after 3 iterations.\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "!python3 /mnt/archgen/users/yilei/tools/hapROH/bam/hapCON_auto.py --bamTable ./Data/BamTable -i test -p 4 --prefix iter_ROH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's run ROH contamination estimates using pileup files as input. First, we generate the pileup files using samtools mpileup.\n",
    "<span style='color:red'>Please change the path to BED files according to your own setup. We have assumed that the test BAM file is in directory \"./Data\"</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools index ./Data/ROH_contam.test.bam\n",
    "!mkdir ./Data/mpileup\n",
    "!for ch in {1..22}; do samtools mpileup --positions /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/1240kChr$ch.bed -r $ch -q 30 -Q 30 -o ./Data/mpileup/test.chr$ch.mpileup ./Data/ROH_contam.test.bam; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style='color:red'>Please make sure that the pileup files are named as \"$iid.chr{1..22}.mpileup\"</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving hdf5 files in ./Data/hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr7.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr5.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr3.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr8.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr1.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr6.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr4.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr9.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr13.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr11.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr14.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr10.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr2.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr15.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr17.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr12.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr19.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr21.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in ./Data/hdf5/test.chr22.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr18.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr16.hdf5\n",
      "saving sample as test in ./Data/hdf5/test.chr20.hdf5\n",
      "finished reading input files, takes 31.771s\n",
      "estimated genotyping error: 0.0019824997768848975\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.510303\t0.635444\n",
      "chr6\t1.042793\t1.131743\n",
      "chr9\t0.96028\t1.029503\n",
      "a total of 3 ROH blocks passing filtering threshold found, total length after trimming: 25.331cM.\n",
      "3 hmm models loaded, takes 185.776s\n",
      "hapsb_multiChunk_preload takes 3.463438034057617\n",
      "hapsb_multiChunk_preload takes 2.803697347640991\n",
      "hapsb_multiChunk_preload takes 2.740445375442505\n",
      "hapsb_multiChunk_preload takes 2.8269646167755127\n",
      "hapsb_multiChunk_preload takes 2.957587242126465\n",
      "hapsb_multiChunk_preload takes 2.716825246810913\n",
      "hapsb_multiChunk_preload takes 2.968498468399048\n",
      "hapsb_multiChunk_preload takes 2.7796630859375\n",
      "hapsb_multiChunk_preload takes 2.751452684402466\n",
      "hapsb_multiChunk_preload takes 2.813168525695801\n",
      "hapsb_multiChunk_preload takes 2.7786178588867188\n",
      "hapsb_multiChunk_preload takes 2.850496768951416\n",
      "hapsb_multiChunk_preload takes 2.6996233463287354\n",
      "hapsb_multiChunk_preload takes 2.766876697540283\n",
      "hapsb_multiChunk_preload takes 2.8286731243133545\n",
      "hapsb_multiChunk_preload takes 2.7798893451690674\n",
      "hapsb_multiChunk_preload takes 2.8072142601013184\n",
      "hapsb_multiChunk_preload takes 2.7777671813964844\n",
      "hapsb_multiChunk_preload takes 2.9509148597717285\n",
      "hapsb_multiChunk_preload takes 2.7456228733062744\n",
      "hapsb_multiChunk_preload takes 2.7214999198913574\n",
      "hapsb_multiChunk_preload takes 2.746959686279297\n",
      "hapsb_multiChunk_preload takes 2.851184129714966\n",
      "estimated contamination rate 0.09847 greater than 0.05, start iteration.\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.472396\t0.639327\n",
      "chr2\t2.220766\t2.279183\n",
      "chr6\t1.042141\t1.195667\n",
      "chr9\t0.96028\t1.054782\n",
      "a total of 4 ROH blocks passing filtering threshold found, total length after trimming: 43.338cM.\n",
      "4 hmm models loaded, takes 270.951s\n",
      "hapsb_multiChunk_preload takes 3.2739884853363037\n",
      "hapsb_multiChunk_preload takes 3.4105966091156006\n",
      "hapsb_multiChunk_preload takes 3.4717438220977783\n",
      "hapsb_multiChunk_preload takes 3.5561699867248535\n",
      "hapsb_multiChunk_preload takes 3.3729047775268555\n",
      "hapsb_multiChunk_preload takes 3.288804054260254\n",
      "hapsb_multiChunk_preload takes 3.4199748039245605\n",
      "hapsb_multiChunk_preload takes 3.3472723960876465\n",
      "hapsb_multiChunk_preload takes 3.444973945617676\n",
      "hapsb_multiChunk_preload takes 3.342738151550293\n",
      "hapsb_multiChunk_preload takes 3.3503260612487793\n",
      "hapsb_multiChunk_preload takes 3.3591015338897705\n",
      "hapsb_multiChunk_preload takes 3.274695873260498\n",
      "hapsb_multiChunk_preload takes 3.494645595550537\n",
      "hapsb_multiChunk_preload takes 3.3410463333129883\n",
      "hapsb_multiChunk_preload takes 3.364915132522583\n",
      "hapsb_multiChunk_preload takes 3.2398622035980225\n",
      "hapsb_multiChunk_preload takes 3.329228401184082\n",
      "hapsb_multiChunk_preload takes 3.3345913887023926\n",
      "hapsb_multiChunk_preload takes 3.3040995597839355\n",
      "hapsb_multiChunk_preload takes 3.291759490966797\n",
      "iteration 2 done, prev contam: 0.09847, current contam: 0.120581\n",
      "Doing Individual test...\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr3/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr5/hmm_run_log.txt\n",
      "Set Output Log path: ./Data/hapRoh_iter/test/chr7/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n",
      "chr1\t0.472343\t0.644013\n",
      "chr2\t2.220766\t2.279213\n",
      "chr6\t1.039623\t1.196152\n",
      "chr9\t0.958626\t1.056485\n",
      "chr19\t0.076272\t0.166039\n",
      "a total of 5 ROH blocks passing filtering threshold found, total length after trimming: 52.427cM.\n",
      "5 hmm models loaded, takes 282.874s\n",
      "hapsb_multiChunk_preload takes 3.234496831893921\n",
      "hapsb_multiChunk_preload takes 3.3449649810791016\n",
      "hapsb_multiChunk_preload takes 3.384476661682129\n",
      "hapsb_multiChunk_preload takes 3.3859078884124756\n",
      "hapsb_multiChunk_preload takes 3.781970262527466\n",
      "hapsb_multiChunk_preload takes 3.3253655433654785\n",
      "hapsb_multiChunk_preload takes 3.4930801391601562\n",
      "hapsb_multiChunk_preload takes 3.4538328647613525\n",
      "hapsb_multiChunk_preload takes 3.465245008468628\n",
      "hapsb_multiChunk_preload takes 3.390801429748535\n",
      "hapsb_multiChunk_preload takes 3.6435933113098145\n",
      "hapsb_multiChunk_preload takes 3.5921452045440674\n",
      "hapsb_multiChunk_preload takes 3.590162754058838\n",
      "hapsb_multiChunk_preload takes 3.4850213527679443\n",
      "hapsb_multiChunk_preload takes 3.4310858249664307\n",
      "hapsb_multiChunk_preload takes 3.357827663421631\n",
      "hapsb_multiChunk_preload takes 3.368096113204956\n",
      "hapsb_multiChunk_preload takes 3.6426007747650146\n",
      "hapsb_multiChunk_preload takes 3.461137056350708\n",
      "hapsb_multiChunk_preload takes 3.698002576828003\n",
      "hapsb_multiChunk_preload takes 3.3097996711730957\n",
      "iteration 3 done, prev contam: 0.120581, current contam: 0.121683\n",
      "contamination rate converged after 3 iterations.\n"
     ]
    }
   ],
   "source": [
    "!python3 /mnt/archgen/users/yilei/tools/hapROH/bam/hapCON_auto.py --mpileup ./Data/mpileup -i test -p 4 --prefix iter_ROH"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
