{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# some tests for hapCON"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test for 1240k panel on a Sardinian"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n"
     ]
    }
   ],
   "source": [
    "# data preparation\n",
    "path2bam=\"../Vignettes/Data/SUA001.bam\"\n",
    "path2bed1240k=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/1240kChrX.bed\"\n",
    "!samtools index $path2bam\n",
    "!samtools mpileup --positions $path2bed1240k -r X -q 30 -Q 30 -o ../Vignettes/Data/SUA001.mpileup $path2bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom_BFGS\n",
    "path2ref1240k=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5\"\n",
    "path2ref1240k_binaryGT = \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/1240k_binary/1240k_binary_chrX.hdf5\"\n",
    "path2meta=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exclude 1033 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 10636\n",
      "number of minor reads at flanking sites: 16\n",
      "number of major reads at focal sites: 1291\n",
      "number of minor reads at focal sites: 27\n",
      "err rate at flanking sites: 0.001502\n",
      "err rate at focal sites: 0.020486\n",
      "saving sample as SUA001 in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "estimated genotyping error by flanking sites: 0.001502\n",
      "number of sites covered by at least one read: 3999, fraction covered: 0.085\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "finished reading mpileup file, takes 1.185.\n",
      "number of sites covered by at least one read: 3999\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: HDF5\n",
      "Loading Individual: SUA001\n",
      "Creating folder /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001/chrX/...\n",
      "\n",
      "Loaded 3999 variants\n",
      "Loaded 1 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "\n",
      "Loaded 47895 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5\n",
      "\n",
      "Intersection on Positions: 3999\n",
      "Nr of Matching Refs: 3999 / 3999\n",
      "Ref/Alt Allele Matching: 3999 / 3999\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 3999 / 3999\n",
      "1843 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "99 / 2504 individuals included in contamination population\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "Extraction of 2772 Haplotypes complete\n",
      "Subset to markers with data: 3999 / 3999\n",
      "Fraction SNPs covered: 1.0000\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.18455\n",
      "Successfully loaded Data from: /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001/chrX/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "estimated contamination rate: 0.101227(0.075982 - 0.126472)\n",
      "nSitesCovered: 3999\n"
     ]
    }
   ],
   "source": [
    "conmle, _, _, nSitesCovered = hapCon_chrom_BFGS(iid=\"SUA001\", mpileup=\"../Vignettes/Data/SUA001.mpileup\",\n",
    "    h5_path1000g = path2ref1240k, meta_path_ref = path2meta, lowmem=False, diploid_ref=True, output=True)\n",
    "assert(conmle > 0.1 and conmle < 0.105)\n",
    "print(f'nSitesCovered: {nSitesCovered}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exclude 1033 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 10636\n",
      "number of minor reads at flanking sites: 16\n",
      "number of major reads at focal sites: 1291\n",
      "number of minor reads at focal sites: 27\n",
      "err rate at flanking sites: 0.001502\n",
      "err rate at focal sites: 0.020486\n",
      "saving sample as SUA001 in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "estimated genotyping error by flanking sites: 0.001502\n",
      "number of sites covered by at least one read: 3999, fraction covered: 0.085\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "finished reading mpileup file, takes 1.088.\n",
      "number of sites covered by at least one read: 3999\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: HDF5\n",
      "Loading Individual: SUA001\n",
      "Creating folder /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001/chrX/...\n",
      "\n",
      "Loaded 3999 variants\n",
      "Loaded 1 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "\n",
      "Loaded 47895 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/1240k_binary/1240k_binary_chrX.hdf5\n",
      "\n",
      "Intersection on Positions: 3999\n",
      "Nr of Matching Refs: 3999 / 3999\n",
      "Ref/Alt Allele Matching: 3999 / 3999\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 3999 / 3999\n",
      "1843 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "99 / 2504 individuals included in contamination population\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Extraction of 2772 reference haplotypes at 3999 sites complete\n",
      "Subset to markers with data: 3999 / 3999\n",
      "Fraction SNPs covered: 1.0000\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.18455\n",
      "Successfully loaded Data from: /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001/chrX/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.056768 mB\n",
      "Total Log likelihood: -1090.992\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1090.992\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1222.487\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1222.487\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1063.607\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1063.607\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1059.594\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1059.594\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1056.763\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1056.763\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.378\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.378\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.206\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.206\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.06496 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.069056 mB\n",
      "Total Log likelihood: -1055.199\n",
      "Minimum Genetic Map: 0.2574 Morgan\n",
      "Maximum Genetic Map: 2.0170 Morgan\n",
      "Gaps bigger than 0.1 cM: 534\n",
      "Maximum Gap: 1.4872 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 257.069056 mB\n",
      "Total Log likelihood: -1055.199\n",
      "estimated contamination rate: 0.101227(0.075982 - 0.126472)\n",
      "nSitesCovered: 3999\n"
     ]
    }
   ],
   "source": [
    "conmle, _, _, nSitesCovered = hapCon_chrom_BFGS(iid=\"SUA001\", mpileup=\"../Vignettes/Data/SUA001.mpileup\",\n",
    "    h5_path1000g = path2ref1240k_binaryGT, meta_path_ref = path2meta, lowmem=True, diploid_ref=True, output=True)\n",
    "assert(conmle > 0.1 and conmle < 0.105)\n",
    "print(f'nSitesCovered: {nSitesCovered}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total number of mapped reads: 14755\n",
      "number of major reads at flanking sites: 10521\n",
      "number of minor reads at flanking sites: 15\n",
      "number of major reads at focal sites: 1291\n",
      "number of minor reads at focal sites: 27\n",
      "err rate at flanking sites: 0.001424\n",
      "err rate at focal sites: 0.020486\n",
      "saving sample as SUA001 in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "estimated genotyping error by flanking sites: 0.001424\n",
      "number of sites covered by at least one read: 3999, fraction covered: 0.085\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "finished reading bam file, takes 89.202.\n",
      "number of sites covered by at least one read: 3999\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/SUA001.hdf5\n",
      "estimated contamination rate: 0.102215(0.076894 - 0.127536)\n"
     ]
    }
   ],
   "source": [
    "# test running directly from bam file\n",
    "conmle, _, _ = hapCon_chrom_BFGS(iid=\"SUA001\", bam=\"../Vignettes/Data/SUA001.bam\",\n",
    "    h5_path1000g = path2ref1240k, meta_path_ref = path2meta)\n",
    "assert(conmle > 0.1 and conmle < 0.105)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# test 1000G panel with XiongNu DA43 sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n"
     ]
    }
   ],
   "source": [
    "path2bam=\"../Vignettes/Data/DA43.bam\"\n",
    "path2bed1kg=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/maf5FilterChrX.bed\"\n",
    "!samtools index $path2bam\n",
    "!samtools mpileup --positions $path2bed1kg -r X -q 30 -Q 30 -o ../Vignettes/Data//DA43.mpileup $path2bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exclude 16544 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 16633\n",
      "number of minor reads at flanking sites: 156\n",
      "number of major reads at focal sites: 1977\n",
      "number of minor reads at focal sites: 41\n",
      "err rate at flanking sites: 0.009292\n",
      "err rate at focal sites: 0.020317\n",
      "saving sample as DA43 in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "estimated genotyping error by flanking sites: 0.009292\n",
      "number of sites covered by at least one read: 17584, fraction covered: 0.085\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "finished reading mpileup file, takes 18.241.\n",
      "number of sites covered by at least one read: 17584\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "in load_data in class HMM_Analyze, ref_states.shape: (1843, 17584)\n",
      "estimated contamination rate: 0.034005(0.027509 - 0.040502)\n",
      "nSitesCovered: 17584\n"
     ]
    }
   ],
   "source": [
    "path2ref1kg=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/maf5_filter_chrX.hdf5\"\n",
    "conmle, _, _, nSitesCovered = hapCon_chrom_BFGS(iid=\"DA43\", mpileup=\"../Vignettes/Data/DA43.mpileup\",\n",
    "    h5_path1000g = path2ref1kg, meta_path_ref = path2meta)\n",
    "assert(conmle > 0.03 and conmle < 0.04)\n",
    "print(f'nSitesCovered: {nSitesCovered}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exclude 16544 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 16633\n",
      "number of minor reads at flanking sites: 156\n",
      "number of major reads at focal sites: 1977\n",
      "number of minor reads at focal sites: 41\n",
      "err rate at flanking sites: 0.009292\n",
      "err rate at focal sites: 0.020317\n",
      "saving sample as DA43 in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "estimated genotyping error by flanking sites: 0.009292\n",
      "number of sites covered by at least one read: 17584, fraction covered: 0.085\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "finished reading mpileup file, takes 18.543.\n",
      "number of sites covered by at least one read: 17584\n",
      "hdf5 file saved to /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Vignettes/Data/DA43.hdf5\n",
      "in load_data in class HMM_Analyze, ref_states.shape: (1843, 2198)\n",
      "estimated contamination rate: 0.034005(0.027509 - 0.040502)\n",
      "nSitesCovered: 17584\n"
     ]
    }
   ],
   "source": [
    "path2ref1kg_binaryGT = \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/maf5_auto/binary/maf5_binary_chrX.hdf5\"\n",
    "conmle, _, _, nSitesCovered = hapCon_chrom_BFGS(iid=\"DA43\", mpileup=\"../Vignettes/Data/DA43.mpileup\",\n",
    "    h5_path1000g = path2ref1kg_binaryGT, meta_path_ref = path2meta, lowmem=True)\n",
    "assert(conmle > 0.03 and conmle < 0.04)\n",
    "print(f'nSitesCovered: {nSitesCovered}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
