{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# a test on Matt's problematic pileup files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hapsburg.PackagesSupport.loadEigenstrat.saveHDF5 import mpileup2hdf5\n",
    "\n",
    "for ch in range(1,23):\n",
    "    mpileup2hdf5(f'./Matt/test.chr{ch}.mpileup',f'/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr{ch}.hdf5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saving hdf5 files in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr1.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr2.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr3.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr4.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr5.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr6.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr7.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr8.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr9.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr10.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr11.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr12.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr13.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr14.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "bug fix for Matt\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr15.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr16.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr17.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr18.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "bug fix for Matt\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr19.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr20.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr21.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "saving sample as test in /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hdf5/test.chr22.hdf5\n",
      "finished reading input files, takes 34.387s\n",
      "estimated genotyping error: 0.023850409557141142\n",
      "fraction of sites covered by at least one read: 0.3062101748853836\n",
      "Doing Individual test...\n",
      "Set Output Log path: /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "number of blocks found with null-model: 4 with total length 33.219cM\n",
      "chr5\t1.285502\t1.415539\n",
      "chr14\t0.252409\t0.335213\n",
      "chr15\t0.619845\t0.707687\n",
      "chr17\t0.659744\t0.731246\n",
      "a total of 4 ROH blocks passing filtering threshold found, total length after trimming: 33.219cM.\n",
      "4 hmm models loaded, takes 170.879s\n",
      "hapsb_multiChunk_preload takes 0.3654348850250244\n",
      "hapsb_multiChunk_preload takes 0.3141770362854004\n",
      "hapsb_multiChunk_preload takes 0.34302186965942383\n",
      "hapsb_multiChunk_preload takes 0.2939779758453369\n",
      "hapsb_multiChunk_preload takes 0.29159092903137207\n",
      "hapsb_multiChunk_preload takes 0.2944674491882324\n",
      "hapsb_multiChunk_preload takes 0.28900909423828125\n",
      "hapsb_multiChunk_preload takes 0.29099178314208984\n",
      "hapsb_multiChunk_preload takes 0.2871828079223633\n",
      "hapsb_multiChunk_preload takes 0.3121342658996582\n",
      "hapsb_multiChunk_preload takes 0.40865278244018555\n",
      "hapsb_multiChunk_preload takes 0.28799891471862793\n",
      "hapsb_multiChunk_preload takes 0.2919182777404785\n",
      "hapsb_multiChunk_preload takes 0.2843773365020752\n",
      "hapsb_multiChunk_preload takes 0.2880425453186035\n",
      "hapsb_multiChunk_preload takes 0.2874443531036377\n",
      "hapsb_multiChunk_preload takes 0.29007911682128906\n",
      "hapsb_multiChunk_preload takes 0.32591748237609863\n",
      "hapsb_multiChunk_preload takes 0.28596043586730957\n",
      "estimated contamination rate 0.052893 greater than 0.05, start iteration.\n",
      "Doing Individual test...\n",
      "Set Output Log path: /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n",
      "chr2\t2.233203\t2.294629\n",
      "chr5\t0.326404\t0.390297\n",
      "chr5\t1.285044\t1.415697\n",
      "chr14\t0.112034\t0.167947\n",
      "chr14\t0.249489\t0.335363\n",
      "chr15\t0.619845\t0.722629\n",
      "chr16\t0.051745\t0.134458\n",
      "chr17\t0.659744\t0.732082\n",
      "a total of 8 ROH blocks passing filtering threshold found, total length after trimming: 57.559cM.\n",
      "8 hmm models loaded, takes 367.324s\n",
      "hapsb_multiChunk_preload takes 0.7066595554351807\n",
      "hapsb_multiChunk_preload takes 0.4553344249725342\n",
      "hapsb_multiChunk_preload takes 0.45717406272888184\n",
      "hapsb_multiChunk_preload takes 0.45864224433898926\n",
      "hapsb_multiChunk_preload takes 0.6327204704284668\n",
      "hapsb_multiChunk_preload takes 0.5433173179626465\n",
      "hapsb_multiChunk_preload takes 0.4655027389526367\n",
      "hapsb_multiChunk_preload takes 0.442979097366333\n",
      "hapsb_multiChunk_preload takes 0.47890806198120117\n",
      "hapsb_multiChunk_preload takes 0.4396226406097412\n",
      "hapsb_multiChunk_preload takes 0.4395933151245117\n",
      "hapsb_multiChunk_preload takes 0.4713122844696045\n",
      "hapsb_multiChunk_preload takes 0.4750840663909912\n",
      "hapsb_multiChunk_preload takes 0.4368302822113037\n",
      "hapsb_multiChunk_preload takes 0.4467449188232422\n",
      "hapsb_multiChunk_preload takes 0.45855236053466797\n",
      "hapsb_multiChunk_preload takes 0.43993091583251953\n",
      "hapsb_multiChunk_preload takes 0.4862079620361328\n",
      "hapsb_multiChunk_preload takes 0.438060998916626\n",
      "iteration 2 done, prev contam: 0.052893, current contam: 0.053902\n",
      "contamination rate converged after 2 iterations.\n",
      "Doing Individual test...\n",
      "Set Output Log path: /mnt/archgen/users/yilei/tools/hapROH/Notebooks/Tests/hapRoh_iter/test/chr1/hmm_run_log.txt\n",
      "Combining Information for 22 Chromosomes...\n"
     ]
    }
   ],
   "source": [
    "!python3 /mnt/archgen/users/yilei/tools/hapROH/bam/hapCON_auto.py --mpileup ./Matt -i test -r /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr --meta /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
