{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# a test on Matt's problematic pileup files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 734\n",
      "number of minor reads at flanking sites: 18\n",
      "number of major reads at focal sites: 15211\n",
      "number of minor reads at focal sites: 1193\n",
      "err rate at flanking sites: 0.023936\n",
      "err rate at focal sites: 0.072726\n",
      "saving sample as test.chr1 in test.chr1.hdf5\n",
      "estimated genotyping error by flanking sites: 0.023936\n",
      "number of sites covered by at least one read: 27889, fraction covered: 0.313\n",
      "hdf5 file saved to test.chr1.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 715\n",
      "number of minor reads at flanking sites: 18\n",
      "number of major reads at focal sites: 14937\n",
      "number of minor reads at focal sites: 1164\n",
      "err rate at flanking sites: 0.024557\n",
      "err rate at focal sites: 0.072294\n",
      "saving sample as test.chr2 in test.chr2.hdf5\n",
      "estimated genotyping error by flanking sites: 0.024557\n",
      "number of sites covered by at least one read: 28220, fraction covered: 0.299\n",
      "hdf5 file saved to test.chr2.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 598\n",
      "number of minor reads at flanking sites: 10\n",
      "number of major reads at focal sites: 11778\n",
      "number of minor reads at focal sites: 955\n",
      "err rate at flanking sites: 0.016447\n",
      "err rate at focal sites: 0.075002\n",
      "saving sample as test.chr3 in test.chr3.hdf5\n",
      "estimated genotyping error by flanking sites: 0.016447\n",
      "number of sites covered by at least one read: 22807, fraction covered: 0.294\n",
      "hdf5 file saved to test.chr3.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 387\n",
      "number of minor reads at flanking sites: 13\n",
      "number of major reads at focal sites: 9166\n",
      "number of minor reads at focal sites: 717\n",
      "err rate at flanking sites: 0.032500\n",
      "err rate at focal sites: 0.072549\n",
      "saving sample as test.chr4 in test.chr4.hdf5\n",
      "estimated genotyping error by flanking sites: 0.032500\n",
      "number of sites covered by at least one read: 19079, fraction covered: 0.277\n",
      "hdf5 file saved to test.chr4.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 748\n",
      "number of minor reads at flanking sites: 20\n",
      "number of major reads at focal sites: 10429\n",
      "number of minor reads at focal sites: 713\n",
      "err rate at flanking sites: 0.026042\n",
      "err rate at focal sites: 0.063992\n",
      "saving sample as test.chr5 in test.chr5.hdf5\n",
      "estimated genotyping error by flanking sites: 0.026042\n",
      "number of sites covered by at least one read: 20384, fraction covered: 0.294\n",
      "hdf5 file saved to test.chr5.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 514\n",
      "number of minor reads at flanking sites: 12\n",
      "number of major reads at focal sites: 11860\n",
      "number of minor reads at focal sites: 788\n",
      "err rate at flanking sites: 0.022814\n",
      "err rate at focal sites: 0.062302\n",
      "saving sample as test.chr6 in test.chr6.hdf5\n",
      "estimated genotyping error by flanking sites: 0.022814\n",
      "number of sites covered by at least one read: 22609, fraction covered: 0.298\n",
      "hdf5 file saved to test.chr6.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 450\n",
      "number of minor reads at flanking sites: 13\n",
      "number of major reads at focal sites: 9579\n",
      "number of minor reads at focal sites: 735\n",
      "err rate at flanking sites: 0.028078\n",
      "err rate at focal sites: 0.071262\n",
      "saving sample as test.chr7 in test.chr7.hdf5\n",
      "estimated genotyping error by flanking sites: 0.028078\n",
      "number of sites covered by at least one read: 18085, fraction covered: 0.302\n",
      "hdf5 file saved to test.chr7.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 458\n",
      "number of minor reads at flanking sites: 9\n",
      "number of major reads at focal sites: 10097\n",
      "number of minor reads at focal sites: 702\n",
      "err rate at flanking sites: 0.019272\n",
      "err rate at focal sites: 0.065006\n",
      "saving sample as test.chr8 in test.chr8.hdf5\n",
      "estimated genotyping error by flanking sites: 0.019272\n",
      "number of sites covered by at least one read: 18650, fraction covered: 0.305\n",
      "hdf5 file saved to test.chr8.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 374\n",
      "number of minor reads at flanking sites: 11\n",
      "number of major reads at focal sites: 8508\n",
      "number of minor reads at focal sites: 615\n",
      "err rate at flanking sites: 0.028571\n",
      "err rate at focal sites: 0.067412\n",
      "saving sample as test.chr9 in test.chr9.hdf5\n",
      "estimated genotyping error by flanking sites: 0.028571\n",
      "number of sites covered by at least one read: 15544, fraction covered: 0.307\n",
      "hdf5 file saved to test.chr9.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 425\n",
      "number of minor reads at flanking sites: 10\n",
      "number of major reads at focal sites: 10164\n",
      "number of minor reads at focal sites: 815\n",
      "err rate at flanking sites: 0.022989\n",
      "err rate at focal sites: 0.074233\n",
      "saving sample as test.chr10 in test.chr10.hdf5\n",
      "estimated genotyping error by flanking sites: 0.022989\n",
      "number of sites covered by at least one read: 18535, fraction covered: 0.315\n",
      "hdf5 file saved to test.chr10.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 410\n",
      "number of minor reads at flanking sites: 13\n",
      "number of major reads at focal sites: 9290\n",
      "number of minor reads at focal sites: 708\n",
      "err rate at flanking sites: 0.030733\n",
      "err rate at focal sites: 0.070814\n",
      "saving sample as test.chr11 in test.chr11.hdf5\n",
      "estimated genotyping error by flanking sites: 0.030733\n",
      "number of sites covered by at least one read: 16922, fraction covered: 0.309\n",
      "hdf5 file saved to test.chr11.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 366\n",
      "number of minor reads at flanking sites: 8\n",
      "number of major reads at focal sites: 8625\n",
      "number of minor reads at focal sites: 643\n",
      "err rate at flanking sites: 0.021390\n",
      "err rate at focal sites: 0.069379\n",
      "saving sample as test.chr12 in test.chr12.hdf5\n",
      "estimated genotyping error by flanking sites: 0.021390\n",
      "number of sites covered by at least one read: 16267, fraction covered: 0.302\n",
      "hdf5 file saved to test.chr12.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 224\n",
      "number of minor reads at flanking sites: 1\n",
      "number of major reads at focal sites: 5598\n",
      "number of minor reads at focal sites: 423\n",
      "err rate at flanking sites: 0.004444\n",
      "err rate at focal sites: 0.070254\n",
      "saving sample as test.chr13 in test.chr13.hdf5\n",
      "estimated genotyping error by flanking sites: 0.004444\n",
      "number of sites covered by at least one read: 10974, fraction covered: 0.281\n",
      "hdf5 file saved to test.chr13.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 285\n",
      "number of minor reads at flanking sites: 4\n",
      "number of major reads at focal sites: 6200\n",
      "number of minor reads at focal sites: 466\n",
      "err rate at flanking sites: 0.013841\n",
      "err rate at focal sites: 0.069907\n",
      "saving sample as test.chr14 in test.chr14.hdf5\n",
      "estimated genotyping error by flanking sites: 0.013841\n",
      "number of sites covered by at least one read: 11086, fraction covered: 0.305\n",
      "hdf5 file saved to test.chr14.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 282\n",
      "number of minor reads at flanking sites: 6\n",
      "number of major reads at focal sites: 6418\n",
      "number of minor reads at focal sites: 514\n",
      "err rate at flanking sites: 0.020833\n",
      "err rate at focal sites: 0.074149\n",
      "saving sample as test.chr15 in test.chr15.hdf5\n",
      "estimated genotyping error by flanking sites: 0.020833\n",
      "number of sites covered by at least one read: 11073, fraction covered: 0.322\n",
      "hdf5 file saved to test.chr15.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 301\n",
      "number of minor reads at flanking sites: 9\n",
      "number of major reads at focal sites: 6781\n",
      "number of minor reads at focal sites: 540\n",
      "err rate at flanking sites: 0.029032\n",
      "err rate at focal sites: 0.073760\n",
      "saving sample as test.chr16 in test.chr16.hdf5\n",
      "estimated genotyping error by flanking sites: 0.029032\n",
      "number of sites covered by at least one read: 11394, fraction covered: 0.331\n",
      "hdf5 file saved to test.chr16.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 307\n",
      "number of minor reads at flanking sites: 5\n",
      "number of major reads at focal sites: 6217\n",
      "number of minor reads at focal sites: 453\n",
      "err rate at flanking sites: 0.016026\n",
      "err rate at focal sites: 0.067916\n",
      "saving sample as test.chr17 in test.chr17.hdf5\n",
      "estimated genotyping error by flanking sites: 0.016026\n",
      "number of sites covered by at least one read: 10141, fraction covered: 0.346\n",
      "hdf5 file saved to test.chr17.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 258\n",
      "number of minor reads at flanking sites: 9\n",
      "number of major reads at focal sites: 5295\n",
      "number of minor reads at focal sites: 397\n",
      "err rate at flanking sites: 0.033708\n",
      "err rate at focal sites: 0.069747\n",
      "saving sample as test.chr18 in test.chr18.hdf5\n",
      "estimated genotyping error by flanking sites: 0.033708\n",
      "number of sites covered by at least one read: 10281, fraction covered: 0.303\n",
      "hdf5 file saved to test.chr18.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 216\n",
      "number of minor reads at flanking sites: 4\n",
      "number of major reads at focal sites: 4310\n",
      "number of minor reads at focal sites: 368\n",
      "err rate at flanking sites: 0.018182\n",
      "err rate at focal sites: 0.078666\n",
      "saving sample as test.chr19 in test.chr19.hdf5\n",
      "estimated genotyping error by flanking sites: 0.018182\n",
      "number of sites covered by at least one read: 6721, fraction covered: 0.364\n",
      "hdf5 file saved to test.chr19.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 226\n",
      "number of minor reads at flanking sites: 8\n",
      "number of major reads at focal sites: 5794\n",
      "number of minor reads at focal sites: 453\n",
      "err rate at flanking sites: 0.034188\n",
      "err rate at focal sites: 0.072515\n",
      "saving sample as test.chr20 in test.chr20.hdf5\n",
      "estimated genotyping error by flanking sites: 0.034188\n",
      "number of sites covered by at least one read: 9634, fraction covered: 0.331\n",
      "hdf5 file saved to test.chr20.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 113\n",
      "number of minor reads at flanking sites: 2\n",
      "number of major reads at focal sites: 2637\n",
      "number of minor reads at focal sites: 226\n",
      "err rate at flanking sites: 0.017391\n",
      "err rate at focal sites: 0.078938\n",
      "saving sample as test.chr21 in test.chr21.hdf5\n",
      "estimated genotyping error by flanking sites: 0.017391\n",
      "number of sites covered by at least one read: 4844, fraction covered: 0.302\n",
      "hdf5 file saved to test.chr21.hdf5\n",
      "exclude 0 sites outside the specified region\n",
      "exclude 0 non-SNP sites\n",
      "number of major reads at flanking sites: 145\n",
      "number of minor reads at flanking sites: 6\n",
      "number of major reads at focal sites: 4025\n",
      "number of minor reads at focal sites: 304\n",
      "err rate at flanking sites: 0.039735\n",
      "err rate at focal sites: 0.070224\n",
      "saving sample as test.chr22 in test.chr22.hdf5\n",
      "estimated genotyping error by flanking sites: 0.039735\n",
      "number of sites covered by at least one read: 6018, fraction covered: 0.381\n",
      "hdf5 file saved to test.chr22.hdf5\n"
     ]
    }
   ],
   "source": [
    "from hapsburg.PackagesSupport.loadEigenstrat.saveHDF5 import mpileup2hdf5\n",
    "\n",
    "for ch in range(1,23):\n",
    "    mpileup2hdf5(f'./Matt/test.chr{ch}.mpileup',f'/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr{ch}.hdf5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/yilei_huang/.local/bin/hapCon_ROH\", line 8, in <module>\n",
      "    sys.exit(main())\n",
      "  File \"/home/yilei_huang/.local/lib/python3.8/site-packages/hapsburg/hapCon_ROH.py\", line 43, in main\n",
      "    set_start_method(\"spawn\")\n",
      "  File \"/usr/lib/python3.8/multiprocessing/context.py\", line 243, in set_start_method\n",
      "    raise RuntimeError('context has already been set')\n",
      "RuntimeError: context has already been set\n"
     ]
    }
   ],
   "source": [
    "!hapCon_ROH --mpileup ./Matt -i test -r /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr --meta /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
