{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")\n",
    "path = \"/mnt/archgen/users/yilei\"   # The Path on Yilei's remote space\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory.\n",
    "\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom  # Need this import\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_ind"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "CPU Count: 40\n",
      "Set path to: /mnt/archgen/users/yilei\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "hapCon_chrom(iid=\"SUC005\", ch=1, save=True, save_fp=False, n_ref=2504, diploid_ref=True, exclude_pops=[], \n",
    "                e_model=\"readcount_contam\", p_model=\"SardHDF5\", readcounts=True,random_allele=False, post_model=\"Standard\", \n",
    "                path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out=\"\",\n",
    "                c=0.05, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0,\n",
    "                max_gap=0, cutoff_post = 0.999, roh_min_l = 0.01, logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 89147 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr1.hdf5\n",
      "\n",
      "Intersection on Positions: 89143\n",
      "Nr of Matching Refs: 89143 / 89143\n",
      "Ref/Alt Allele Matching: 89079 / 89143\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 89079 / 89143\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 89079)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 89079)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 60218 / 89079\n",
      "Fraction SNPs covered: 0.6760\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.44575\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t1\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0201 Morgan\n",
      "Maximum Genetic Map: 2.8627 Morgan\n",
      "Gaps bigger than 0.1 cM: 329\n",
      "Maximum Gap: 3.1359 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 60218)\n",
      "Loaded Observations:\n",
      "(2, 60218)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2954.743808 mB\n",
      "Total Log likelihood: -35783.111\n",
      "Total Log likelihood: -35783.111\n",
      "hapcon_chr returns result: -35783.111073376574\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "-35783.111073376574"
      ]
     },
     "metadata": {},
     "execution_count": 2
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "hapCon_ind(\"SUC005\", chs=range(1,6), \n",
    "            path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "            h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "            meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "            folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out=\"\",\n",
    "            e_model=\"readcount_contam\", p_model=\"SardHDF5\", post_model=\"Standard\",\n",
    "            processes=1, delete=False, output=False, save=True, save_fp=False, \n",
    "            n_ref=2504, diploid_ref=True, exclude_pops=[], readcounts=True, random_allele=False,\n",
    "            c=0.15, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.00, \n",
    "            cutoff_post = 0.999, max_gap=0, roh_min_l = 0.01, logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 89147 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr1.hdf5\n",
      "\n",
      "Intersection on Positions: 89143\n",
      "Nr of Matching Refs: 89143 / 89143\n",
      "Ref/Alt Allele Matching: 89079 / 89143\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 89079 / 89143\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 89079)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 89079)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 60218 / 89079\n",
      "Fraction SNPs covered: 0.6760\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.44575\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t1\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.15\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0201 Morgan\n",
      "Maximum Genetic Map: 2.8627 Morgan\n",
      "Gaps bigger than 0.1 cM: 329\n",
      "Maximum Gap: 3.1359 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 60218)\n",
      "Loaded Observations:\n",
      "(2, 60218)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2970.39872 mB\n",
      "Total Log likelihood: -35619.888\n",
      "Total Log likelihood: -35619.888\n",
      "hapcon_chr returns result: -35619.88754316375\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 94239 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr2.hdf5\n",
      "\n",
      "Intersection on Positions: 94237\n",
      "Nr of Matching Refs: 94205 / 94237\n",
      "Ref/Alt Allele Matching: 94137 / 94237\n",
      "Flipped Ref/Alt Alleles for 8 SNPs\n",
      "Together: 94145 / 94237\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 94145)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 8 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 94145)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 61791 / 94145\n",
      "Fraction SNPs covered: 0.6563\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr2/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.35986\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr2/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t2\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.15\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0001 Morgan\n",
      "Maximum Genetic Map: 2.6883 Morgan\n",
      "Gaps bigger than 0.1 cM: 315\n",
      "Maximum Gap: 3.9385 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 61791)\n",
      "Loaded Observations:\n",
      "(2, 61791)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 3041.325056 mB\n",
      "Total Log likelihood: -36001.420\n",
      "Total Log likelihood: -36001.420\n",
      "hapcon_chr returns result: -36001.42018400625\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77650\n",
      "Nr of Matching Refs: 77606 / 77650\n",
      "Ref/Alt Allele Matching: 77553 / 77650\n",
      "Flipped Ref/Alt Alleles for 10 SNPs\n",
      "Together: 77563 / 77650\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77563)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 10 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77563)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 50550 / 77563\n",
      "Fraction SNPs covered: 0.6517\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr3/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.34172\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr3/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.15\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 299\n",
      "Maximum Gap: 0.3910 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 50550)\n",
      "Loaded Observations:\n",
      "(2, 50550)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2534.621184 mB\n",
      "Total Log likelihood: -29814.438\n",
      "Total Log likelihood: -29814.438\n",
      "hapcon_chr returns result: -29814.438072916928\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 68759 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr4.hdf5\n",
      "\n",
      "Intersection on Positions: 68759\n",
      "Nr of Matching Refs: 68709 / 68759\n",
      "Ref/Alt Allele Matching: 68663 / 68759\n",
      "Flipped Ref/Alt Alleles for 11 SNPs\n",
      "Together: 68674 / 68759\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 68674)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 11 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 68674)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 42540 / 68674\n",
      "Fraction SNPs covered: 0.6194\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr4/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.25068\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr4/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t4\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.15\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0034 Morgan\n",
      "Maximum Genetic Map: 2.1454 Morgan\n",
      "Gaps bigger than 0.1 cM: 290\n",
      "Maximum Gap: 2.0440 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 42540)\n",
      "Loaded Observations:\n",
      "(2, 42540)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2173.534208 mB\n",
      "Total Log likelihood: -24404.443\n",
      "Total Log likelihood: -24404.443\n",
      "hapcon_chr returns result: -24404.442789576504\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 69399 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr5.hdf5\n",
      "\n",
      "Intersection on Positions: 69394\n",
      "Nr of Matching Refs: 69307 / 69394\n",
      "Ref/Alt Allele Matching: 69252 / 69394\n",
      "Flipped Ref/Alt Alleles for 16 SNPs\n",
      "Together: 69268 / 69394\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 69268)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 16 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 69268)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 44914 / 69268\n",
      "Fraction SNPs covered: 0.6484\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr5/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.35466\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr5/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t5\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.15\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0004 Morgan\n",
      "Maximum Genetic Map: 2.0409 Morgan\n",
      "Gaps bigger than 0.1 cM: 276\n",
      "Maximum Gap: 1.1070 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 44914)\n",
      "Loaded Observations:\n",
      "(2, 44914)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2280.574976 mB\n",
      "Total Log likelihood: -26338.833\n",
      "Total Log likelihood: -26338.833\n",
      "hapcon_chr returns result: -26338.832553237142\n",
      "total log likelihood: -152179.0211429006\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "-152179.0211429006"
      ]
     },
     "metadata": {},
     "execution_count": 5
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "import multiprocessing as mul\n",
    "\n",
    "def f(x, y):\n",
    "    return x*y\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    with mul.Pool(5) as pool:\n",
    "        rel = pool.starmap(f,[[1,2],[3,4],[5,6],[7,8],[9,10]])\n",
    "        print(rel)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[2, 12, 30, 56, 90]\n"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}