{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")\n",
    "path = \"/mnt/archgen/users/yilei\"   # The Path on Yilei's remote space\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory.\n",
    "\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom  # Need this import\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_ind"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "CPU Count: 40\nSet path to: /mnt/archgen/users/yilei\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "hapCon_chrom(iid=\"SUC005\", ch=1, save=True, save_fp=False, n_ref=2504, diploid_ref=True, exclude_pops=[], \n",
    "                e_model=\"readcount_contam\", p_model=\"SardHDF5\", readcounts=True,random_allele=False, post_model=\"Standard\", \n",
    "                path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out=\"\",\n",
    "                c=0.05, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0,\n",
    "                max_gap=0, cutoff_post = 0.999, roh_min_l = 0.01, logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 89147 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr1.hdf5\n",
      "\n",
      "Intersection on Positions: 89143\n",
      "Nr of Matching Refs: 89143 / 89143\n",
      "Ref/Alt Allele Matching: 89079 / 89143\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 89079 / 89143\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 89079)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 89079)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 60218 / 89079\n",
      "Fraction SNPs covered: 0.6760\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.44575\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr1/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t1\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0201 Morgan\n",
      "Maximum Genetic Map: 2.8627 Morgan\n",
      "Gaps bigger than 0.1 cM: 329\n",
      "Maximum Gap: 3.1359 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 60218)\n",
      "Loaded Observations:\n",
      "(2, 60218)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2954.743808 mB\n",
      "Total Log likelihood: -35783.111\n",
      "Total Log likelihood: -35783.111\n",
      "hapcon_chr returns result: -35783.111073376574\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "-35783.111073376574"
      ]
     },
     "metadata": {},
     "execution_count": 2
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "hapCon_ind(\"SUC005\", chs=range(20,23), \n",
    "            path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "            h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "            meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "            folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out=\"\",\n",
    "            e_model=\"readcount_contam\", p_model=\"SardHDF5\", post_model=\"Standard\",\n",
    "            processes=3, delete=False, output=False, save=True, save_fp=False, \n",
    "            n_ref=2504, diploid_ref=True, exclude_pops=[], readcounts=True, random_allele=False,\n",
    "            c=0.05, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.00, \n",
    "            cutoff_post = 0.999, max_gap=0, roh_min_l = 0.01, logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Running 3 total jobs; 3 in parallel.\n",
      "Starting Pool of multiple workers...\n",
      "Using Rescaled HMM.Using Rescaled HMM.Using Rescaled HMM.\n",
      "\n",
      "\n",
      "Loaded Pre Processing Model: SardHDF5Loaded Pre Processing Model: SardHDF5Loaded Pre Processing Model: SardHDF5\n",
      "\n",
      "\n",
      "Loading Individual: SUC005Loading Individual: SUC005Loading Individual: SUC005\n",
      "\n",
      "\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 1145647 variants\n",
      "Loaded 1145647 variants\n",
      "\n",
      "\n",
      "Loaded 3039 individualsLoaded 3039 individualsLoaded 3039 individuals\n",
      "\n",
      "\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 16038 variants\n",
      "Loaded 15802 variants\n",
      "\n",
      "\n",
      "Loaded 2504 individualsLoaded 2504 individualsLoaded 2504 individuals\n",
      "\n",
      "\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr20.hdf5HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr22.hdf5HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr21.hdf5\n",
      "\n",
      "\n",
      "\n",
      "Intersection on Positions: 15802\n",
      "Nr of Matching Refs: 15704 / 15802\n",
      "Ref/Alt Allele Matching: 15684 / 15802\n",
      "Flipped Ref/Alt Alleles for 21 SNPs\n",
      "Intersection on Positions: 16036\n",
      "\n",
      "Together: 15705 / 15802Nr of Matching Refs: 15949 / 16036\n",
      "\n",
      "Ref/Alt Allele Matching: 15930 / 16036\n",
      "Flipped Ref/Alt Alleles for 18 SNPs\n",
      "Together: 15948 / 16036\n",
      "\n",
      "Intersection on Positions: 29077\n",
      "2504 / 2504 Individuals included in ReferenceNr of Matching Refs: 28908 / 29077\n",
      "\n",
      "Ref/Alt Allele Matching: 28864 / 29077Extracting up to 2504 Individuals\n",
      "\n",
      "2504 / 2504 Individuals included in ReferenceFlipped Ref/Alt Alleles for 43 SNPs\n",
      "\n",
      "Extracting up to 2504 IndividualsTogether: 28907 / 29077\n",
      "\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 15705)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 28907)\n",
      "Exctraction of hdf5 done. Subsetting...!Extraction of 2 Haplotypes complete\n",
      "\n",
      "size of gts matrix: (2, 15948)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 21 SNPs...\n",
      "Flipping Ref/Alt Allele in target for 43 SNPs...\n",
      "Flipping Ref/Alt Allele in target for 18 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 15705)\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 15948)\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 28907)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 11848 / 15705\n",
      "Fraction SNPs covered: 0.7544\n",
      "Subset to markers with data: 10511 / 15948\n",
      "Fraction SNPs covered: 0.6591\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr22/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.77338\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr21/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.43802\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr22/\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr21/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t22\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0169 Morgan\n",
      "Maximum Genetic Map: 0.7408 Morgan\n",
      "Gaps bigger than 0.1 cM: 92\n",
      "Maximum Gap: 1.6949 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t21\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0089 Morgan\n",
      "Maximum Genetic Map: 0.6277 Morgan\n",
      "Gaps bigger than 0.1 cM: 102\n",
      "Maximum Gap: 0.5653 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 20607 / 28907\n",
      "Fraction SNPs covered: 0.7129\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr20/\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.55607\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr20/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 160\n",
      "Maximum Gap: 0.6809 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 10511)\n",
      "Loaded Observations:\n",
      "(2, 10511)\n",
      "Reference Number: 5008\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 11848)\n",
      "Loaded Observations:\n",
      "(2, 11848)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 736.41984 mB\n",
      "Total Log likelihood: -7581.730\n",
      "Total Log likelihood: -7581.730\n",
      "hapcon_chr returns result: -7581.730432295715\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 676.192256 mB\n",
      "Total Log likelihood: -6797.848\n",
      "Total Log likelihood: -6797.848\n",
      "hapcon_chr returns result: -6797.84840291573\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 20607)\n",
      "Loaded Observations:\n",
      "(2, 20607)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1132.990464 mB\n",
      "Total Log likelihood: -13148.607\n",
      "Total Log likelihood: -13148.607\n",
      "hapcon_chr returns result: -13148.607398623777\n",
      "results of multiprocess: None\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ]
}