{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "This notebook is to try a few functions of hapROH to get started. Also to play around with a few file formats.\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "\n",
    "###use local version and not  pip installed version\n",
    "#sys.path.append(\"/mnt/archgen/users/yilei/tools/hapROH/package\") # Append local Hapsburg Folder\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]\n",
    "\n",
    "### Fill in your own path here!\n",
    "path = \"/mnt/archgen/users/yilei\"   # The Path on Yilei's remote space\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory.\n"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "CPU Count: 40\n",
      "Set path to: /mnt/archgen/users/yilei\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "The first is to run hapROH on a normal autosome. Should be straightforward.\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(17,18), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29314\n",
      "Nr of Matching Refs: 29314 / 29314\n",
      "Ref/Alt Matching: 29292 / 29314\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29292 / 29314\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 22918 / 29292\n",
      "Fraction SNPs covered: 0.7824\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 22918)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0003 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 152\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 22918)\n",
      "Loaded Observations:\n",
      "(2, 22918)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1264.914432 mB\n",
      "Total Log likelihood: -9070.009\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Fraction Markers in ROH: 0.1732\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.03 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "The following is to try to run hapROH on chrX. chrX is by definition \"ROH\" everywhere for male (Villabruna is a male individual). \n",
    "\n",
    "1. The first run uses the default transition model.\n",
    "\n",
    "\n",
    "2. The second run uses a transition model that prohibits transition out of ROH state. A hack is to set the transition rate out of ROH to be 0. (also need to set initial prob accordingly?)\n",
    "\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(23,24), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/chX1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/chX1240/chr23.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Matching: 47766 / 47895\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 47766 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 23276 / 47766\n",
      "Fraction SNPs covered: 0.4873\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 23276)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t23\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/chX1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.2085 Morgan\n",
      "Maximum Genetic Map: 2.0206 Morgan\n",
      "Gaps bigger than 0.1 cM: 322\n",
      "Maximum Gap: 1.1490 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(3776, 23276)\n",
      "Loaded Observations:\n",
      "(2, 23276)\n",
      "Reference Number: 3775\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1041.682432 mB\n",
      "Total Log likelihood: -2037.192\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Fraction Markers in ROH: 0.9939\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=5 ROH Blocks > 1.0 cM\n",
      "Longest Block: 98.39 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {
    "tags": []
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(23,24), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/chX1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='chrX_tran', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=0, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/chX1240/chr23.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Matching: 47766 / 47895\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 47766 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 23276 / 47766\n",
      "Fraction SNPs covered: 0.4873\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 23276)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t23\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/chX1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tchrX_tran\n",
      "roh_in\t1\n",
      "roh_out\t0\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.2085 Morgan\n",
      "Maximum Genetic Map: 2.0206 Morgan\n",
      "Gaps bigger than 0.1 cM: 322\n",
      "Maximum Gap: 1.1490 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(3776, 23276)\n",
      "Loaded Observations:\n",
      "(2, 23276)\n",
      "Reference Number: 3775\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1032.368128 mB\n",
      "Total Log likelihood: -2002.640\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Fraction Markers in ROH: 0.9979\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=1 ROH Blocks > 1.0 cM\n",
      "Longest Block: 179.97 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now let's use readcount data."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "hapsb_ind(iid=\"SUC005\", chs=range(17,18), \n",
    "          path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran_SA_rc_', \n",
    "          e_model='readcount', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=True, random_allele=False, \n",
    "          roh_in=1, roh_out=10, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual SUC005...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "\n",
      "Intersection on Positions: 29313\n",
      "Nr of Matching Refs: 29159 / 29313\n",
      "Ref/Alt Allele Matching: 29124 / 29313\n",
      "Flipped Ref/Alt Alleles for 37 SNPs\n",
      "Together: 29161 / 29313\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 29161)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 37 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 29161)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 21019 / 29161\n",
      "Fraction SNPs covered: 0.7208\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.61578\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Loaded Emission Model: readcount\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_rc_\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0002 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 152\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 21019)\n",
      "Loaded Observations:\n",
      "(2, 21019)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1180.024832 mB\n",
      "Total Log likelihood: -14136.773\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Fraction Markers in ROH: 0.1347\n",
      "Called n=0 ROH Blocks > 1.0 cM\n",
      "Longest Block: nan cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "hapsb_ind(iid=\"SUC001\", chs=range(1, 23), \n",
    "          path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran_SA_haploid', \n",
    "          e_model='haploid', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=10, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=True, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual SUC001...\n",
      "Running 22 total jobs; 1 in parallel.\n",
      "Starting Pool of multiple workers...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 89147 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr1.hdf5\n",
      "\n",
      "Intersection on Positions: 89143\n",
      "Nr of Matching Refs: 89143 / 89143\n",
      "Ref/Alt Allele Matching: 89079 / 89143\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 89079 / 89143\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 89079)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 89079)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 57717 / 89079\n",
      "Fraction SNPs covered: 0.6479\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr1/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr1/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t1\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0201 Morgan\n",
      "Maximum Genetic Map: 2.8627 Morgan\n",
      "Gaps bigger than 0.1 cM: 354\n",
      "Maximum Gap: 2.9245 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 57717)\n",
      "Loaded Observations:\n",
      "(2, 57717)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2812.366848 mB\n",
      "Total Log likelihood: -21431.672\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr1/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr1/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2394\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=11 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.74 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr1/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 94239 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr2.hdf5\n",
      "\n",
      "Intersection on Positions: 94237\n",
      "Nr of Matching Refs: 94205 / 94237\n",
      "Ref/Alt Allele Matching: 94137 / 94237\n",
      "Flipped Ref/Alt Alleles for 8 SNPs\n",
      "Together: 94145 / 94237\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 94145)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 8 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 94145)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 59123 / 94145\n",
      "Fraction SNPs covered: 0.6280\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr2/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr2/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t2\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0001 Morgan\n",
      "Maximum Genetic Map: 2.6883 Morgan\n",
      "Gaps bigger than 0.1 cM: 321\n",
      "Maximum Gap: 2.9097 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 59123)\n",
      "Loaded Observations:\n",
      "(2, 59123)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2879.131648 mB\n",
      "Total Log likelihood: -21430.639\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr2/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr2/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2861\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=16 ROH Blocks > 1.0 cM\n",
      "Longest Block: 3.15 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr2/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77650\n",
      "Nr of Matching Refs: 77606 / 77650\n",
      "Ref/Alt Allele Matching: 77553 / 77650\n",
      "Flipped Ref/Alt Alleles for 10 SNPs\n",
      "Together: 77563 / 77650\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77563)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 10 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77563)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 48504 / 77563\n",
      "Fraction SNPs covered: 0.6253\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr3/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr3/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 294\n",
      "Maximum Gap: 0.4131 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 48504)\n",
      "Loaded Observations:\n",
      "(2, 48504)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2408.972288 mB\n",
      "Total Log likelihood: -18468.931\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr3/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr3/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2235\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=7 ROH Blocks > 1.0 cM\n",
      "Longest Block: 3.04 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr3/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 68759 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr4.hdf5\n",
      "\n",
      "Intersection on Positions: 68759\n",
      "Nr of Matching Refs: 68709 / 68759\n",
      "Ref/Alt Allele Matching: 68663 / 68759\n",
      "Flipped Ref/Alt Alleles for 11 SNPs\n",
      "Together: 68674 / 68759\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 68674)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 11 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 68674)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 40955 / 68674\n",
      "Fraction SNPs covered: 0.5964\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr4/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr4/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t4\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0034 Morgan\n",
      "Maximum Genetic Map: 2.1454 Morgan\n",
      "Gaps bigger than 0.1 cM: 292\n",
      "Maximum Gap: 2.0513 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 40955)\n",
      "Loaded Observations:\n",
      "(2, 40955)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2068.836352 mB\n",
      "Total Log likelihood: -15444.225\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr4/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr4/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2272\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=7 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.84 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr4/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 69399 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr5.hdf5\n",
      "\n",
      "Intersection on Positions: 69394\n",
      "Nr of Matching Refs: 69307 / 69394\n",
      "Ref/Alt Allele Matching: 69252 / 69394\n",
      "Flipped Ref/Alt Alleles for 16 SNPs\n",
      "Together: 69268 / 69394\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 69268)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 16 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 69268)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 43002 / 69268\n",
      "Fraction SNPs covered: 0.6208\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr5/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr5/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t5\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0004 Morgan\n",
      "Maximum Genetic Map: 2.0409 Morgan\n",
      "Gaps bigger than 0.1 cM: 280\n",
      "Maximum Gap: 1.1070 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 43002)\n",
      "Loaded Observations:\n",
      "(2, 43002)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2161.278976 mB\n",
      "Total Log likelihood: -16419.668\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr5/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr5/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2176\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=7 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.36 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr5/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 75870 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr6.hdf5\n",
      "\n",
      "Intersection on Positions: 75866\n",
      "Nr of Matching Refs: 75743 / 75866\n",
      "Ref/Alt Allele Matching: 75675 / 75866\n",
      "Flipped Ref/Alt Alleles for 33 SNPs\n",
      "Together: 75708 / 75866\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 75708)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 33 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 75708)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 47409 / 75708\n",
      "Fraction SNPs covered: 0.6262\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr6/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr6/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t6\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0035 Morgan\n",
      "Maximum Genetic Map: 1.9203 Morgan\n",
      "Gaps bigger than 0.1 cM: 235\n",
      "Maximum Gap: 0.5090 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 47409)\n",
      "Loaded Observations:\n",
      "(2, 47409)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2360.393728 mB\n",
      "Total Log likelihood: -17548.611\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr6/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr6/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2602\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=5 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.69 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr6/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 59890 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr7.hdf5\n",
      "\n",
      "Intersection on Positions: 59889\n",
      "Nr of Matching Refs: 59781 / 59889\n",
      "Ref/Alt Allele Matching: 59715 / 59889\n",
      "Flipped Ref/Alt Alleles for 31 SNPs\n",
      "Together: 59746 / 59889\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 59746)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 31 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 59746)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 37077 / 59746\n",
      "Fraction SNPs covered: 0.6206\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr7/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr7/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t7\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.8715 Morgan\n",
      "Gaps bigger than 0.1 cM: 273\n",
      "Maximum Gap: 0.8416 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 37077)\n",
      "Loaded Observations:\n",
      "(2, 37077)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1894.641664 mB\n",
      "Total Log likelihood: -14385.436\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr7/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr7/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2306\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=6 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.82 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr7/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 61133 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr8.hdf5\n",
      "\n",
      "Intersection on Positions: 61130\n",
      "Nr of Matching Refs: 61022 / 61130\n",
      "Ref/Alt Allele Matching: 60960 / 61130\n",
      "Flipped Ref/Alt Alleles for 27 SNPs\n",
      "Together: 60987 / 61130\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 60987)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 27 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 60987)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 38333 / 60987\n",
      "Fraction SNPs covered: 0.6285\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr8/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr8/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t8\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.6800 Morgan\n",
      "Gaps bigger than 0.1 cM: 198\n",
      "Maximum Gap: 1.8932 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 38333)\n",
      "Loaded Observations:\n",
      "(2, 38333)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1951.268864 mB\n",
      "Total Log likelihood: -14671.750\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr8/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr8/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2449\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=9 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.81 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr8/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 50668 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr9.hdf5\n",
      "\n",
      "Intersection on Positions: 50668\n",
      "Nr of Matching Refs: 50539 / 50668\n",
      "Ref/Alt Allele Matching: 50493 / 50668\n",
      "Flipped Ref/Alt Alleles for 28 SNPs\n",
      "Together: 50521 / 50668\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 50521)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 28 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 50521)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 31971 / 50521\n",
      "Fraction SNPs covered: 0.6328\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr9/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr9/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t9\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0017 Morgan\n",
      "Maximum Genetic Map: 1.6630 Morgan\n",
      "Gaps bigger than 0.1 cM: 234\n",
      "Maximum Gap: 2.3088 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 31971)\n",
      "Loaded Observations:\n",
      "(2, 31971)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1665.294336 mB\n",
      "Total Log likelihood: -11816.356\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr9/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr9/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2726\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=6 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.63 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr9/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 58804 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr10.hdf5\n",
      "\n",
      "Intersection on Positions: 58803\n",
      "Nr of Matching Refs: 58657 / 58803\n",
      "Ref/Alt Allele Matching: 58591 / 58803\n",
      "Flipped Ref/Alt Alleles for 30 SNPs\n",
      "Together: 58621 / 58803\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 58621)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 30 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 58621)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 38013 / 58621\n",
      "Fraction SNPs covered: 0.6485\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr10/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr10/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t10\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0001 Morgan\n",
      "Maximum Genetic Map: 1.8091 Morgan\n",
      "Gaps bigger than 0.1 cM: 244\n",
      "Maximum Gap: 2.0340 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 38013)\n",
      "Loaded Observations:\n",
      "(2, 38013)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1937.67424 mB\n",
      "Total Log likelihood: -14207.495\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr10/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr10/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2463\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=6 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.63 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr10/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 54833 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr11.hdf5\n",
      "\n",
      "Intersection on Positions: 54831\n",
      "Nr of Matching Refs: 54669 / 54831\n",
      "Ref/Alt Allele Matching: 54616 / 54831\n",
      "Flipped Ref/Alt Alleles for 33 SNPs\n",
      "Together: 54649 / 54831\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 54649)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 33 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 54649)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 34698 / 54649\n",
      "Fraction SNPs covered: 0.6349\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr11/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr11/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t11\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.5822 Morgan\n",
      "Gaps bigger than 0.1 cM: 215\n",
      "Maximum Gap: 0.5672 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 34698)\n",
      "Loaded Observations:\n",
      "(2, 34698)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1788.301312 mB\n",
      "Total Log likelihood: -13016.822\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr11/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr11/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2502\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=6 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.04 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr11/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 53911 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr12.hdf5\n",
      "\n",
      "Intersection on Positions: 53911\n",
      "Nr of Matching Refs: 53713 / 53911\n",
      "Ref/Alt Allele Matching: 53672 / 53911\n",
      "Flipped Ref/Alt Alleles for 38 SNPs\n",
      "Together: 53710 / 53911\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 53710)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 38 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 53710)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 33764 / 53710\n",
      "Fraction SNPs covered: 0.6286\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr12/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr12/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t12\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0017 Morgan\n",
      "Maximum Genetic Map: 1.7468 Morgan\n",
      "Gaps bigger than 0.1 cM: 230\n",
      "Maximum Gap: 0.8437 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 33764)\n",
      "Loaded Observations:\n",
      "(2, 33764)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1746.259968 mB\n",
      "Total Log likelihood: -12563.367\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr12/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr12/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2742\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=9 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.84 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr12/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 39018 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr13.hdf5\n",
      "\n",
      "Intersection on Positions: 39015\n",
      "Nr of Matching Refs: 38885 / 39015\n",
      "Ref/Alt Allele Matching: 38841 / 39015\n",
      "Flipped Ref/Alt Alleles for 30 SNPs\n",
      "Together: 38871 / 39015\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 38871)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 30 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 38871)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 23559 / 38871\n",
      "Fraction SNPs covered: 0.6061\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr13/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr13/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t13\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0021 Morgan\n",
      "Maximum Genetic Map: 1.2570 Morgan\n",
      "Gaps bigger than 0.1 cM: 185\n",
      "Maximum Gap: 1.2216 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 23559)\n",
      "Loaded Observations:\n",
      "(2, 23559)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1286.32832 mB\n",
      "Total Log likelihood: -9027.037\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr13/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr13/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2186\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 3.27 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr13/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 36301 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr14.hdf5\n",
      "\n",
      "Intersection on Positions: 36301\n",
      "Nr of Matching Refs: 36166 / 36301\n",
      "Ref/Alt Allele Matching: 36126 / 36301\n",
      "Flipped Ref/Alt Alleles for 33 SNPs\n",
      "Together: 36159 / 36301\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 36159)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 33 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 36159)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 22658 / 36159\n",
      "Fraction SNPs covered: 0.6266\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr14/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr14/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t14\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0161 Morgan\n",
      "Maximum Genetic Map: 1.2020 Morgan\n",
      "Gaps bigger than 0.1 cM: 168\n",
      "Maximum Gap: 2.8973 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 22658)\n",
      "Loaded Observations:\n",
      "(2, 22658)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1245.708288 mB\n",
      "Total Log likelihood: -8551.630\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr14/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr14/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2515\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.61 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr14/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 34429 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr15.hdf5\n",
      "\n",
      "Intersection on Positions: 34425\n",
      "Nr of Matching Refs: 34285 / 34425\n",
      "Ref/Alt Allele Matching: 34249 / 34425\n",
      "Flipped Ref/Alt Alleles for 24 SNPs\n",
      "Together: 34273 / 34425\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 34273)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 24 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 34273)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 22442 / 34273\n",
      "Fraction SNPs covered: 0.6548\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr15/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr15/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t15\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.4134 Morgan\n",
      "Gaps bigger than 0.1 cM: 166\n",
      "Maximum Gap: 11.4957 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 22442)\n",
      "Loaded Observations:\n",
      "(2, 22442)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1235.972096 mB\n",
      "Total Log likelihood: -8244.191\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr15/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr15/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2806\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=6 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.13 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr15/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 34402 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr16.hdf5\n",
      "\n",
      "Intersection on Positions: 34399\n",
      "Nr of Matching Refs: 34254 / 34399\n",
      "Ref/Alt Allele Matching: 34201 / 34399\n",
      "Flipped Ref/Alt Alleles for 30 SNPs\n",
      "Together: 34231 / 34399\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 34231)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 30 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 34231)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 23035 / 34231\n",
      "Fraction SNPs covered: 0.6729\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr16/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr16/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t16\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0000 Morgan\n",
      "Maximum Genetic Map: 1.3403 Morgan\n",
      "Gaps bigger than 0.1 cM: 189\n",
      "Maximum Gap: 0.8458 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 23035)\n",
      "Loaded Observations:\n",
      "(2, 23035)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1262.743552 mB\n",
      "Total Log likelihood: -8978.467\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr16/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr16/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2045\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.02 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr16/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "\n",
      "Intersection on Positions: 29313\n",
      "Nr of Matching Refs: 29159 / 29313\n",
      "Ref/Alt Allele Matching: 29124 / 29313\n",
      "Flipped Ref/Alt Alleles for 37 SNPs\n",
      "Together: 29161 / 29313\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 29161)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 37 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 29161)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 19860 / 29161\n",
      "Fraction SNPs covered: 0.6810\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr17/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr17/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0002 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 170\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 19860)\n",
      "Loaded Observations:\n",
      "(2, 19860)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1119.62112 mB\n",
      "Total Log likelihood: -7268.437\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr17/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr17/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.3017\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=8 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.40 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr17/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 33917 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr18.hdf5\n",
      "\n",
      "Intersection on Positions: 33916\n",
      "Nr of Matching Refs: 33774 / 33916\n",
      "Ref/Alt Allele Matching: 33746 / 33916\n",
      "Flipped Ref/Alt Alleles for 43 SNPs\n",
      "Together: 33789 / 33916\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 33789)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 43 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 33789)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 20921 / 33789\n",
      "Fraction SNPs covered: 0.6192\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr18/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr18/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t18\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0025 Morgan\n",
      "Maximum Genetic Map: 1.1771 Morgan\n",
      "Gaps bigger than 0.1 cM: 181\n",
      "Maximum Gap: 0.4827 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 20921)\n",
      "Loaded Observations:\n",
      "(2, 20921)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1167.568896 mB\n",
      "Total Log likelihood: -7922.491\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr18/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr18/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2339\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=5 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.05 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr18/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 18460 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr19.hdf5\n",
      "\n",
      "Intersection on Positions: 18459\n",
      "Nr of Matching Refs: 18381 / 18459\n",
      "Ref/Alt Allele Matching: 18340 / 18459\n",
      "Flipped Ref/Alt Alleles for 23 SNPs\n",
      "Together: 18363 / 18459\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 18363)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 23 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 18363)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 12802 / 18363\n",
      "Fraction SNPs covered: 0.6972\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr19/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr19/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t19\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0773 Morgan\n",
      "Gaps bigger than 0.1 cM: 171\n",
      "Maximum Gap: 0.6534 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 12802)\n",
      "Loaded Observations:\n",
      "(2, 12802)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 801.562624 mB\n",
      "Total Log likelihood: -5188.666\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr19/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr19/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.1618\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=2 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.65 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr19/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 29078 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr20.hdf5\n",
      "\n",
      "Intersection on Positions: 29077\n",
      "Nr of Matching Refs: 28908 / 29077\n",
      "Ref/Alt Allele Matching: 28864 / 29077\n",
      "Flipped Ref/Alt Alleles for 43 SNPs\n",
      "Together: 28907 / 29077\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 28907)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 43 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 28907)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 19679 / 28907\n",
      "Fraction SNPs covered: 0.6808\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr20/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr20/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t20\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 1.0821 Morgan\n",
      "Gaps bigger than 0.1 cM: 166\n",
      "Maximum Gap: 0.5887 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 19679)\n",
      "Loaded Observations:\n",
      "(2, 19679)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1111.584768 mB\n",
      "Total Log likelihood: -7465.910\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr20/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr20/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2443\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=3 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.86 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr20/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 16038 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr21.hdf5\n",
      "\n",
      "Intersection on Positions: 16036\n",
      "Nr of Matching Refs: 15949 / 16036\n",
      "Ref/Alt Allele Matching: 15930 / 16036\n",
      "Flipped Ref/Alt Alleles for 18 SNPs\n",
      "Together: 15948 / 16036\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 15948)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 18 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 15948)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 10096 / 15948\n",
      "Fraction SNPs covered: 0.6331\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr21/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr21/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t21\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0085 Morgan\n",
      "Maximum Genetic Map: 0.6277 Morgan\n",
      "Gaps bigger than 0.1 cM: 103\n",
      "Maximum Gap: 1.7139 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 10096)\n",
      "Loaded Observations:\n",
      "(2, 10096)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 679.583744 mB\n",
      "Total Log likelihood: -4007.056\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr21/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr21/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2168\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=3 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.81 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr21/default_tran_SA_haploid/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC001\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 15802 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr22.hdf5\n",
      "\n",
      "Intersection on Positions: 15802\n",
      "Nr of Matching Refs: 15704 / 15802\n",
      "Ref/Alt Allele Matching: 15684 / 15802\n",
      "Flipped Ref/Alt Alleles for 21 SNPs\n",
      "Together: 15705 / 15802\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 15705)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 21 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 15705)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 11177 / 15705\n",
      "Fraction SNPs covered: 0.7117\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr22/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr22/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC001\n",
      "ch\t22\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0178 Morgan\n",
      "Maximum Genetic Map: 0.7408 Morgan\n",
      "Gaps bigger than 0.1 cM: 101\n",
      "Maximum Gap: 1.6277 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 11177)\n",
      "Loaded Observations:\n",
      "(2, 11177)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 728.313856 mB\n",
      "Total Log likelihood: -4344.601\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr22/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr22/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.2664\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.30 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC001/chr22/default_tran_SA_haploid/roh.csv\n",
      "Combining Information for 22 Chromosomes...\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [],
   "outputs": [],
   "metadata": {}
  }
 ]
}