{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4,
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "This notebook is to try a few functions of hapROH to get started. Also to play around with a few file formats.\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "\n",
    "###use local version and not  pip installed version\n",
    "#sys.path.append(\"/mnt/archgen/users/yilei/tools/hapROH/package\") # Append local Hapsburg Folder\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]\n",
    "\n",
    "### Fill in your own path here!\n",
    "path = \"/mnt/archgen/users/yilei\"   # The Path on Yilei's remote space\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory.\n"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "CPU Count: 40\nSet path to: /mnt/archgen/users/yilei\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "source": [
    "The first is to run hapROH on a normal autosome. Should be straightforward.\n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 29314\n",
      "Nr of Matching Refs: 29314 / 29314\n",
      "Ref/Alt Matching: 29292 / 29314\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 29292 / 29314\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 22918 / 29292\n",
      "Fraction SNPs covered: 0.7824\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 22918)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\t\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0003 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 152\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 22918)\n",
      "Loaded Observations:\n",
      "(2, 22918)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1264.914432 mB\n",
      "Total Log likelihood: -9070.009\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/\n",
      "Fraction Markers in ROH: 0.1732\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=4 ROH Blocks > 1.0 cM\n",
      "Longest Block: 2.03 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr17/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(17,18), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "The following is to try to run hapROH on chrX. chrX is by definition \"ROH\" everywhere for male (Villabruna is a male individual). \n",
    "\n",
    "1. The first run uses the default transition model.\n",
    "\n",
    "\n",
    "2. The second run uses a transition model that prohibits transition out of ROH state. A hack is to set the transition rate out of ROH to be 0. (also need to set initial prob accordingly?)\n",
    "\n"
   ],
   "metadata": {}
  },
  {
   "source": [],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(23,24), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/chX1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/chX1240/chr23.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Matching: 47766 / 47895\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 47766 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 23276 / 47766\n",
      "Fraction SNPs covered: 0.4873\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 23276)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t23\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/chX1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.2085 Morgan\n",
      "Maximum Genetic Map: 2.0206 Morgan\n",
      "Gaps bigger than 0.1 cM: 322\n",
      "Maximum Gap: 1.1490 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(3776, 23276)\n",
      "Loaded Observations:\n",
      "(2, 23276)\n",
      "Reference Number: 3775\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1041.682432 mB\n",
      "Total Log likelihood: -2037.192\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran\n",
      "Fraction Markers in ROH: 0.9939\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=5 ROH Blocks > 1.0 cM\n",
      "Longest Block: 98.39 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/default_tran/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {
    "tags": []
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "hapsb_ind(iid=\"I6696\", chs=range(23,24), \n",
    "          path_targets='./Data/1240K/v44.3_1240K_public', # The path before the .ind, .snp, .geno\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/chX1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='chrX_tran', \n",
    "          e_model='haploid', p_model='Eigenstrat', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=0, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual I6696...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: Eigenstrat\n",
      "Loading Individual: I6696\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/chX1240/chr23.hdf5\n",
      "Eigenstrat packed: True\n",
      "3 Eigenstrat Files with 9275 Individuals and 1233013 SNPs\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Matching: 47766 / 47895\n",
      "Flipped Ref/Alt Matching: 0\n",
      "Together: 47766 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Reduced to markers with data: 23276 / 47766\n",
      "Fraction SNPs covered: 0.4873\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 23276)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Flipping Ref/Alt Alleles in target for 0 SNPs...\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tI6696\n",
      "ch\t23\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tEigenstrat\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t./Data/1240K/v44.3_1240K_public\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/chX1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tchrX_tran\n",
      "roh_in\t1\n",
      "roh_out\t0\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.2085 Morgan\n",
      "Maximum Genetic Map: 2.0206 Morgan\n",
      "Gaps bigger than 0.1 cM: 322\n",
      "Maximum Gap: 1.1490 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(3776, 23276)\n",
      "Loaded Observations:\n",
      "(2, 23276)\n",
      "Reference Number: 3775\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1032.368128 mB\n",
      "Total Log likelihood: -2002.640\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran\n",
      "Fraction Markers in ROH: 0.9979\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=1 ROH Blocks > 1.0 cM\n",
      "Longest Block: 179.97 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/I6696/chr23/chrX_tran/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "source": [
    "Now let's use readcount data."
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual SUC005...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "\n",
      "Intersection on Positions: 29313\n",
      "Nr of Matching Refs: 29159 / 29313\n",
      "Ref/Alt Allele Matching: 29124 / 29313\n",
      "Flipped Ref/Alt Alleles for 37 SNPs\n",
      "Together: 29161 / 29313\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 29161)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 37 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 29161)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 21019 / 29161\n",
      "Fraction SNPs covered: 0.7208\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 2.61578\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Loaded Emission Model: readcount\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_rc_\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0002 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 152\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 21019)\n",
      "Loaded Observations:\n",
      "(2, 21019)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1180.024832 mB\n",
      "Total Log likelihood: -14136.773\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_\n",
      "Fraction Markers in ROH: 0.1347\n",
      "Called n=0 ROH Blocks > 1.0 cM\n",
      "Longest Block: nan cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_rc_/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ],
   "source": [
    "hapsb_ind(iid=\"SUC005\", chs=range(17,18), \n",
    "          path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran_SA_rc_', \n",
    "          e_model='readcount', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=True, random_allele=False, \n",
    "          roh_in=1, roh_out=10, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ]
  },
  {
   "source": [
    "hapsb_ind(iid=\"SUC005\", chs=range(17,18), \n",
    "          path_targets='/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5',\n",
    "          h5_path1000g='./Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='./Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./tools/hapROH/test_output/contaminate/first_try/', prefix_out='default_tran_SA_haploid', \n",
    "          e_model='haploid', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=10, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "cell_type": "code",
   "metadata": {},
   "execution_count": 4,
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual SUC005...\n",
      "Running 1 total jobs; 1 in parallel.\n",
      "Running single process...\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: SUC005\n",
      "\n",
      "Loaded 1145647 variants\n",
      "Loaded 3039 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "\n",
      "Loaded 29314 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ./Data/1000G/1000g1240khdf5/all1240/chr17.hdf5\n",
      "\n",
      "Intersection on Positions: 29313\n",
      "Nr of Matching Refs: 29159 / 29313\n",
      "Ref/Alt Allele Matching: 29124 / 29313\n",
      "Flipped Ref/Alt Alleles for 37 SNPs\n",
      "Together: 29161 / 29313\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 29161)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 37 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 29161)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 21019 / 29161\n",
      "Fraction SNPs covered: 0.7208\n",
      "Successfully saved target individual data to: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_haploid\n",
      "Shuffling phase of target...\n",
      "Successfully loaded Data from: ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_haploid\n",
      "Loaded Emission Model: haploid\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapsb_chrom:\n",
      "iid\tSUC005\n",
      "ch\t17\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\thaploid\n",
      "p_model\tSardHDF5\n",
      "readcounts\tFalse\n",
      "random_allele\tTrue\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/SA_1240KHDF5/marcus2020.h5\n",
      "h5_path1000g\t./Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t./Data/1000G/meta_df_all.csv\n",
      "folder_out\t./tools/hapROH/test_output/contaminate/first_try/\n",
      "prefix_out\tdefault_tran_SA_haploid\n",
      "roh_in\t1\n",
      "roh_out\t10\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: -0.0002 Morgan\n",
      "Maximum Genetic Map: 1.2847 Morgan\n",
      "Gaps bigger than 0.1 cM: 152\n",
      "Maximum Gap: 1.2341 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 21019)\n",
      "Loaded Observations:\n",
      "(2, 21019)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 1211.1872 mB\n",
      "Total Log likelihood: -8357.134\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_haploid.\n",
      "Successfully loaded for PP. from ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_haploid\n",
      "Fraction Markers in ROH: 0.1921\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=3 ROH Blocks > 1.0 cM\n",
      "Longest Block: 1.19 cM\n",
      "Successfully saved to ./tools/hapROH/test_output/contaminate/first_try/SUC005/chr17/default_tran_SA_haploid/roh.csv\n",
      "Run finished successfully!\n"
     ]
    }
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}