{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "analyze He's data to see how much ROH there is.\n",
    "Turns out there isn't much.\n",
    "In the two minimally contaminated male samples, there are only about 25cM of ROH genome-wide, I don't think that's enough for reliable inference.\n",
    "I think the Carribean dataset is a better alternative."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "iid=\"GFW002.003\""
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "# convert mpileup to hdf5\n",
    "from toHDF5 import mpileup2hdf5\n",
    "\n",
    "basepath=f\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}\"\n",
    "for ch in range(1, 23):\n",
    "    print(f'processing chr{ch}')\n",
    "    mpileup2hdf5(f'{basepath}/mpileup/{iid}.chr{ch}.mpileup', \n",
    "        f'/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr{ch}.hdf5',\n",
    "        iid=iid, outPath=f'{basepath}/hdf5/')\n"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "processing chr1\n",
      "number of major reads at flanking sites: 1856166.0\n",
      "number of minor reads at flanking sites: 10789.0\n",
      "number of major reads at focal sites: 113261.0\n",
      "number of minor reads at focal sites: 8654.0\n",
      "err rate at flanking sites: 0.005778928790463616\n",
      "err rate at focal sites: 0.07098388221301727\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr1.hdf5\n",
      "processing chr2\n",
      "number of major reads at flanking sites: 1763038.0\n",
      "number of minor reads at flanking sites: 8313.0\n",
      "number of major reads at focal sites: 124623.0\n",
      "number of minor reads at focal sites: 9271.0\n",
      "err rate at flanking sites: 0.004693028089859096\n",
      "err rate at focal sites: 0.06924134016460783\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr2.hdf5\n",
      "processing chr3\n",
      "number of major reads at flanking sites: 1395233.0\n",
      "number of minor reads at flanking sites: 5010.0\n",
      "number of major reads at focal sites: 123298.0\n",
      "number of minor reads at focal sites: 8815.0\n",
      "err rate at flanking sites: 0.003577950398609384\n",
      "err rate at focal sites: 0.06672318394102018\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr3.hdf5\n",
      "processing chr4\n",
      "number of major reads at flanking sites: 1093698.0\n",
      "number of minor reads at flanking sites: 3914.0\n",
      "number of major reads at focal sites: 94662.0\n",
      "number of minor reads at focal sites: 6254.0\n",
      "err rate at flanking sites: 0.0035659231130854984\n",
      "err rate at focal sites: 0.061972333425819495\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr4.hdf5\n",
      "processing chr5\n",
      "number of major reads at flanking sites: 1252406.0\n",
      "number of minor reads at flanking sites: 4865.0\n",
      "number of major reads at focal sites: 108718.0\n",
      "number of minor reads at focal sites: 8043.0\n",
      "err rate at flanking sites: 0.0038694919392875523\n",
      "err rate at focal sites: 0.06888430212142753\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr5.hdf5\n",
      "processing chr6\n",
      "number of major reads at flanking sites: 1416747.0\n",
      "number of minor reads at flanking sites: 4783.0\n",
      "number of major reads at focal sites: 134101.0\n",
      "number of minor reads at focal sites: 10763.0\n",
      "err rate at flanking sites: 0.0033646845300486098\n",
      "err rate at focal sites: 0.07429727192401149\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr6.hdf5\n",
      "processing chr7\n",
      "number of major reads at flanking sites: 1092057.0\n",
      "number of minor reads at flanking sites: 2703.0\n",
      "number of major reads at focal sites: 115005.0\n",
      "number of minor reads at focal sites: 9318.0\n",
      "err rate at flanking sites: 0.0024690343088896198\n",
      "err rate at focal sites: 0.07494992881445911\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr7.hdf5\n",
      "processing chr8\n",
      "number of major reads at flanking sites: 1122457.0\n",
      "number of minor reads at flanking sites: 2682.0\n",
      "number of major reads at focal sites: 121861.0\n",
      "number of minor reads at focal sites: 9487.0\n",
      "err rate at flanking sites: 0.0023837054799451447\n",
      "err rate at focal sites: 0.07222797454091422\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr8.hdf5\n",
      "processing chr9\n",
      "number of major reads at flanking sites: 957095.0\n",
      "number of minor reads at flanking sites: 2053.0\n",
      "number of major reads at focal sites: 104196.0\n",
      "number of minor reads at focal sites: 7946.0\n",
      "err rate at flanking sites: 0.0021404413083278076\n",
      "err rate at focal sites: 0.07085659253446523\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr9.hdf5\n",
      "processing chr10\n",
      "number of major reads at flanking sites: 1153621.0\n",
      "number of minor reads at flanking sites: 2595.0\n",
      "number of major reads at focal sites: 126620.0\n",
      "number of minor reads at focal sites: 10375.0\n",
      "err rate at flanking sites: 0.002244390321531617\n",
      "err rate at focal sites: 0.07573269097412315\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr10.hdf5\n",
      "processing chr11\n",
      "number of major reads at flanking sites: 1061238.0\n",
      "number of minor reads at flanking sites: 2195.0\n",
      "number of major reads at focal sites: 117469.0\n",
      "number of minor reads at focal sites: 9748.0\n",
      "err rate at flanking sites: 0.00206406985677518\n",
      "err rate at focal sites: 0.07662497936596524\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr11.hdf5\n",
      "processing chr12\n",
      "number of major reads at flanking sites: 997329.0\n",
      "number of minor reads at flanking sites: 2183.0\n",
      "number of major reads at focal sites: 108401.0\n",
      "number of minor reads at focal sites: 8781.0\n",
      "err rate at flanking sites: 0.0021840658241221718\n",
      "err rate at focal sites: 0.0749347169360482\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr12.hdf5\n",
      "processing chr13\n",
      "number of major reads at flanking sites: 637728.0\n",
      "number of minor reads at flanking sites: 1020.0\n",
      "number of major reads at focal sites: 74503.0\n",
      "number of minor reads at focal sites: 5861.0\n",
      "err rate at flanking sites: 0.0015968738845366247\n",
      "err rate at focal sites: 0.07293066547210193\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr13.hdf5\n",
      "processing chr14\n",
      "number of major reads at flanking sites: 671642.0\n",
      "number of minor reads at flanking sites: 1086.0\n",
      "number of major reads at focal sites: 78732.0\n",
      "number of minor reads at focal sites: 6103.0\n",
      "err rate at flanking sites: 0.0016143225791107254\n",
      "err rate at focal sites: 0.07193964755112867\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr14.hdf5\n",
      "processing chr15\n",
      "number of major reads at flanking sites: 693083.0\n",
      "number of minor reads at flanking sites: 1153.0\n",
      "number of major reads at focal sites: 80475.0\n",
      "number of minor reads at focal sites: 7077.0\n",
      "err rate at flanking sites: 0.0016608185112843472\n",
      "err rate at focal sites: 0.08083196271929824\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr15.hdf5\n",
      "processing chr16\n",
      "number of major reads at flanking sites: 757533.0\n",
      "number of minor reads at flanking sites: 1768.0\n",
      "number of major reads at focal sites: 83568.0\n",
      "number of minor reads at focal sites: 5778.0\n",
      "err rate at flanking sites: 0.0023284573574906394\n",
      "err rate at focal sites: 0.06466993485998254\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr16.hdf5\n",
      "processing chr17\n",
      "number of major reads at flanking sites: 688475.0\n",
      "number of minor reads at flanking sites: 1785.0\n",
      "number of major reads at focal sites: 73290.0\n",
      "number of minor reads at focal sites: 5946.0\n",
      "err rate at flanking sites: 0.0025859820937038217\n",
      "err rate at focal sites: 0.07504164773587763\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr17.hdf5\n",
      "processing chr18\n",
      "number of major reads at flanking sites: 607125.0\n",
      "number of minor reads at flanking sites: 1261.0\n",
      "number of major reads at focal sites: 66169.0\n",
      "number of minor reads at focal sites: 4989.0\n",
      "err rate at flanking sites: 0.002072697267852975\n",
      "err rate at focal sites: 0.07011158267517356\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr18.hdf5\n",
      "processing chr19\n",
      "number of major reads at flanking sites: 484716.0\n",
      "number of minor reads at flanking sites: 1543.0\n",
      "number of major reads at focal sites: 50273.0\n",
      "number of minor reads at focal sites: 4299.0\n",
      "err rate at flanking sites: 0.003173206048628406\n",
      "err rate at focal sites: 0.07877666202448141\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr19.hdf5\n",
      "processing chr20\n",
      "number of major reads at flanking sites: 647780.0\n",
      "number of minor reads at flanking sites: 1583.0\n",
      "number of major reads at focal sites: 67491.0\n",
      "number of minor reads at focal sites: 5104.0\n",
      "err rate at flanking sites: 0.00243777363354549\n",
      "err rate at focal sites: 0.07030787244300572\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr20.hdf5\n",
      "processing chr21\n",
      "number of major reads at flanking sites: 298644.0\n",
      "number of minor reads at flanking sites: 549.0\n",
      "number of major reads at focal sites: 34827.0\n",
      "number of minor reads at focal sites: 2957.0\n",
      "err rate at flanking sites: 0.0018349359777802287\n",
      "err rate at focal sites: 0.07826063942409485\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr21.hdf5\n",
      "processing chr22\n",
      "number of major reads at flanking sites: 415215.0\n",
      "number of minor reads at flanking sites: 748.0\n",
      "number of major reads at focal sites: 48400.0\n",
      "number of minor reads at focal sites: 4040.0\n",
      "err rate at flanking sites: 0.0017982368624132434\n",
      "err rate at focal sites: 0.07704042715484363\n",
      "saving sample as GFW002.003 in /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr22.hdf5\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_chunk_negloglik"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "hapsb_ind(iid, chs=range(1,23), \n",
    "    path_targets_prefix = f\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hdf5\",\n",
    "    h5_path1000g = \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "    meta_path_ref = \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\",\n",
    "    folder_out=f\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hapRoh/\", prefix_out=\"\",\n",
    "    e_model=\"readcount\", p_model=\"SardHDF5\", post_model=\"Standard\",\n",
    "    processes=4, delete=False, output=True, save=True, save_fp=False, \n",
    "    n_ref=2504, diploid_ref=True, exclude_pops=[], readcounts=True, random_allele=False,\n",
    "    roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=1e-3, \n",
    "    cutoff_post = 0.999, max_gap=0, roh_min_l = 0.04, logfile=True, combine=True, \n",
    "    file_result=\"_roh_full.csv\")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Doing Individual GFW002.003...\n",
      "Running 22 total jobs; 4 in parallel.\n",
      "Starting Pool of multiple workers...\n",
      "Set Output Log path: /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hapRoh/GFW002.003/chr1/hmm_run_log.txtSet Output Log path: /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hapRoh/GFW002.003/chr5/hmm_run_log.txtSet Output Log path: /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hapRoh/GFW002.003/chr3/hmm_run_log.txtSet Output Log path: /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hapRoh/GFW002.003/chr7/hmm_run_log.txt\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "output_type": "error",
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_2853723/1021984554.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m hapsb_ind(iid, chs=range(1,23), \n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mpath_targets_prefix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hdf5\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mh5_path1000g\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mmeta_path_ref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mfolder_out\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34mf\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hapRoh/\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprefix_out\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/archgen/users/yilei/tools/hapROH/package/hapsburg/PackagesSupport/hapsburg_run.py\u001b[0m in \u001b[0;36mhapsb_ind\u001b[0;34m(iid, chs, path_targets_prefix, path_targets, h5_path1000g, meta_path_ref, folder_out, prefix_out, e_model, p_model, post_model, processes, delete, output, save, save_fp, n_ref, diploid_ref, exclude_pops, readcounts, random_allele, c, roh_in, roh_out, roh_jump, e_rate, e_rate_ref, cutoff_post, max_gap, roh_min_l, logfile, combine, file_result)\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m     \u001b[0;31m### Run the analysis in parallel\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mmulti_run\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhapsb_chrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprocesses\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocesses\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;31m### Merge results for that Individual\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/archgen/users/yilei/tools/hapROH/package/hapsburg/PackagesSupport/parallel_runs/helper_functions.py\u001b[0m in \u001b[0;36mmulti_run\u001b[0;34m(fun, prms, processes)\u001b[0m\n\u001b[1;32m     37\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Starting Pool of multiple workers...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mmp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocesses\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocesses\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m             \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstarmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfun\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Running single process...\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.8/multiprocessing/pool.py\u001b[0m in \u001b[0;36mstarmap\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    370\u001b[0m         \u001b[0;31m`\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0mbecomes\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m         '''\n\u001b[0;32m--> 372\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_map_async\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0miterable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstarmapstar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchunksize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    373\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    374\u001b[0m     def starmap_async(self, func, iterable, chunksize=None, callback=None,\n",
      "\u001b[0;32m/usr/lib/python3.8/multiprocessing/pool.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 765\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    766\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mready\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    767\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mTimeoutError\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.8/multiprocessing/pool.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    760\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    761\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 762\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_event\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    763\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.8/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    556\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    557\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 558\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    559\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    560\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.8/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    300\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    301\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 302\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    303\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    304\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "hapsb_chunk_negloglik(iid, 3, 1000, -1000, \n",
    "    f\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hdf5/{iid}.chr3.hdf5\",\n",
    "    \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\", \n",
    "    \"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\",\n",
    "    f\"/mnt/archgen/users/yilei/Data/He/contamX/{iid}/hapRoh/\",\n",
    "    0.0, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0,\n",
    "    save=False, save_fp=False, n_ref=2504, diploid_ref=True, \n",
    "    exclude_pops=[], e_model=\"readcount_contam\", p_model=\"SardHDF5\", \n",
    "    readcounts=True, random_allele=False,\n",
    "    post_model=\"Standard\", prefix_out=\"\", logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: GFW002.003\n",
      "\n",
      "Loaded 58677 variants\n",
      "Loaded 1 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hdf5/GFW002.003.chr3.hdf5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "subsetting reference genotype... starting from 1000th marker until -1000th marker.\n",
      "\n",
      "Intersection on Positions: 58677\n",
      "Nr of Matching Refs: 58677 / 58677\n",
      "Ref/Alt Allele Matching: 58677 / 58677\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 58677 / 58677\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "including all individuals from the reference panel as the contamination population\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 58677)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 58677)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "size of gts_con matrix: (58677, 2504, 2)\n",
      "size of gts_con matrix: (5008, 58677)\n",
      "Subset to markers with data: 45788 / 58677\n",
      "Fraction SNPs covered: 0.7803\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 3.16994\n",
      "Successfully loaded Data from: /mnt/archgen/users/yilei/Data/He/contamX/GFW002.003/hapRoh/GFW002.003/chr3/\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "Minimum Genetic Map: 0.1273 Morgan\n",
      "Maximum Genetic Map: 1.6643 Morgan\n",
      "Gaps bigger than 0.1 cM: 177\n",
      "Maximum Gap: 0.2348 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 5008\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "26043.443786579654"
      ]
     },
     "metadata": {},
     "execution_count": 9
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "# compare results to make sure my code for hapsb_chunk_negloglik is correct"
   ],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.8.10",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}