{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "play with .bam file with pysam "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import os\n",
    "basePath = \"/mnt/archgen/users/yilei/Data/iberian_BAM\"\n",
    "id = \"I10280\"\n",
    "path2bam = f\"{basePath}/{id}_1240k.bam\"\n",
    "#path2bam = basePath + f\"/{id}/{id}_1240k.bam\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf = pysam.AlignmentFile(path2bam, \"rb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of major reads at flanking sites: 19899.0\n",
      "number of minor reads at flanking sites: 28.0\n",
      "number of major reads at focal sites: 2482.0\n",
      "number of minor reads at focal sites: 6.0\n",
      "err rate at flanking sites: 0.0014051287198273699\n",
      "err rate at focal sites: 0.002411575562700965\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import gzip\n",
    "import time\n",
    "# compare mismatch rate at variable and non-variable sites\n",
    "# grep all variable sites on chrX from 1240k panel\n",
    "f = h5py.File('/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5', 'r')\n",
    "pos = f['variants/POS']\n",
    "rec = f['variants/MAP']\n",
    "ref = f['variants/REF'][:]\n",
    "alt = f['variants/ALT'][:, 0]\n",
    "assert(len(ref) == len(pos))\n",
    "assert(len(alt) == len(pos))\n",
    "# 1 here means there is just one sample\n",
    "# 2 : [ref read count, alt read count]\n",
    "ad = np.zeros((len(pos), 1, 2))\n",
    "\n",
    "#use ANGSD's HapMapChrX instead?\n",
    "# pos = []\n",
    "# with gzip.open('/mnt/archgen/users/yilei/bin/angsd/RES/HapMapChrX.gz') as file:\n",
    "#     for line in file:\n",
    "#         phy, *_ = line.strip().split()\n",
    "#         pos.append(int(phy))\n",
    "# print(f'number of SNP sites to check: {len(pos)}')\n",
    "pos = np.unique(np.array(pos))\n",
    "pos_pysam = pos-1 # pysam's use 0-based coordinate system\n",
    "\n",
    "major_adj = 0\n",
    "minor_adj = 0\n",
    "major_foc = 0\n",
    "minor_foc = 0\n",
    "base2index = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "flip = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "for i, p in enumerate(pos_pysam):\n",
    "    #print(f'###### focal site: {p} #######')\n",
    "    for pileupcolumn in sf.pileup('X', p-4, p+5, min_mapping_quality=30, min_base_quality=20, truncate=True):\n",
    "        basePos = pileupcolumn.pos\n",
    "        cov = 0\n",
    "        rc = np.zeros(4)\n",
    "        #print(f'coverage at base {pileupcolumn.pos}: {pileupcolumn.n}')\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            if not pileupread.is_del and not pileupread.is_refskip:\n",
    "                baseCall = pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "                #if pileupread.alignment.is_reverse: # do we need this ?!\n",
    "                #    baseCall = flip[baseCall]\n",
    "                rc[base2index[baseCall]] += 1\n",
    "                cov += 1\n",
    "                # if basePos == p:\n",
    "                #     if baseCall == ref[i]:\n",
    "                #         ad[i, 0, 0] += 1\n",
    "                #     elif baseCall == alt[i]:\n",
    "                #         ad[i, 0, 1] += 1\n",
    "                #     else:\n",
    "                #         if (baseCall == 'T' and 'C' in [ref[i], alt[i]]) or (baseCall == 'A' and 'G' in [ref[i], alt[i]]):\n",
    "                #             print(f'potential DNA damage pattern')\n",
    "                #         print(f'at position {pos[i]} the read {baseCall} is neither ref {ref[i]} nor alt {alt[i]}.')\n",
    "\n",
    "        if cov == 1:\n",
    "            continue # sites covered by only 1 read is not informative for estimating genotyping error\n",
    "        if basePos != p:\n",
    "            major_adj += np.max(rc)\n",
    "            minor_adj += np.sum(rc) - np.max(rc)\n",
    "        else:\n",
    "            major_foc += np.max(rc)\n",
    "            minor_foc += np.sum(rc) - np.max(rc)\n",
    "        if np.sum(rc == 0) < 2:\n",
    "            print(f'more than 2 genotype calls at {basePos}')\n",
    "\n",
    "print(f'number of major reads at flanking sites: {major_adj}')\n",
    "print(f'number of minor reads at flanking sites: {minor_adj}')\n",
    "print(f'number of major reads at focal sites: {major_foc}')\n",
    "print(f'number of minor reads at focal sites: {minor_foc}')\n",
    "\n",
    "print(f'err rate at flanking sites: {minor_adj/(minor_adj + major_adj)}')\n",
    "print(f'err rate at focal sites: {minor_foc/(minor_foc + major_foc)}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run hap\bCon on the hdf5 file made from bam\n",
    "# compare this result to ANGSD estimate, does it make sense?\n",
    "import os\n",
    "\n",
    "l = len(pos)\n",
    "k = 1 # 1 bam = 1 sample\n",
    "rec = f['variants/MAP']\n",
    "gt = np.zeros((l, k, 2)) # dummy gt matrix, won't be used for inference\n",
    "dt = h5py.special_dtype(vlen=str)  # To have no problem with saving\n",
    "\n",
    "path2save = f'{basePath}/{id}/{id}.hdf5'\n",
    "if os.path.exists(path2save):  # Do a Deletion of existing File there\n",
    "    os.remove(path2save)\n",
    "\n",
    "\n",
    "with h5py.File(f'{basePath}/{id}/{id}.hdf5', 'w') as f0:\n",
    "    # Create all the Groups\n",
    "    f_map = f0.create_dataset(\"variants/MAP\", (l,), dtype='f')\n",
    "    f_ad = f0.create_dataset(\"calldata/AD\", (l, k, 2), dtype='i')\n",
    "    f_ref = f0.create_dataset(\"variants/REF\", (l,), dtype=dt)\n",
    "    f_alt = f0.create_dataset(\"variants/ALT\", (l,), dtype=dt)\n",
    "    f_pos = f0.create_dataset(\"variants/POS\", (l,), dtype='i')\n",
    "    f_gt = f0.create_dataset(\"calldata/GT\", (l, k, 2), dtype='i')\n",
    "    f_samples = f0.create_dataset(\"samples\", (k,), dtype=dt)\n",
    "\n",
    "    # Save the Data\n",
    "    f_map[:] = rec\n",
    "    f_ad[:] = ad\n",
    "    f_ref[:] = ref.astype(\"S1\")\n",
    "    f_alt[:] = alt.astype(\"S1\")\n",
    "    f_pos[:] = pos\n",
    "    f_gt[:] = gt\n",
    "    f_samples[:] = np.array([id]).astype(\"S10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom  # Need this import\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom_2d  # Need this import\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: I10939\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 1 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/iberian_BAM/I10939/I10939.hdf5\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Allele Matching: 47895 / 47895\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 47895 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 47895)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 47895)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Subset to markers with data: 6857 / 47895\n",
      "Fraction SNPs covered: 0.1432\n",
      "Successfully saved target individual data to: ./test_output/contaminate/first_try/I10939/chrX/fast_\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.21613\n",
      "Successfully loaded Data from: ./test_output/contaminate/first_try/I10939/chrX/fast_\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tI10939\n",
      "ch\tX\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "conPop\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/iberian_BAM/I10939/I10939.hdf5\n",
      "h5_path1000g\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\n",
      "folder_out\t./test_output/contaminate/first_try/\n",
      "prefix_out\tfast_\n",
      "c\t[0.    0.005 0.01  0.015 0.02  0.025 0.03  0.035 0.04  0.045 0.05  0.055\n",
      " 0.06  0.065 0.07  0.075 0.08  0.085 0.09  0.095 0.1   0.105 0.11  0.115\n",
      " 0.12  0.125 0.13  0.135 0.14  0.145 0.15  0.155 0.16  0.165 0.17  0.175\n",
      " 0.18  0.185 0.19  0.195]\n",
      "roh_in\t1\n",
      "roh_out\t0\n",
      "roh_jump\t300\n",
      "e_rate\t0.002\n",
      "e_rate_ref\t0.001\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "([-1243.4409089360374,\n",
       "  -1233.4201624139791,\n",
       "  -1230.461253467486,\n",
       "  -1230.4580372272158,\n",
       "  -1232.1560409018175,\n",
       "  -1234.983206877424,\n",
       "  -1238.6251637593614,\n",
       "  -1242.8886343347826,\n",
       "  -1247.6454285136272,\n",
       "  -1252.8057197738444,\n",
       "  -1258.3038772593322,\n",
       "  -1264.0903443494144,\n",
       "  -1270.1266931010155,\n",
       "  -1276.3824635621506,\n",
       "  -1282.8330622583221,\n",
       "  -1289.4583179758497,\n",
       "  -1296.2414610453611,\n",
       "  -1303.1683843488206,\n",
       "  -1310.2270969734157,\n",
       "  -1317.4073128040525,\n",
       "  -1324.700135652764,\n",
       "  -1332.0978147595197,\n",
       "  -1339.5935524575607,\n",
       "  -1347.181351093628,\n",
       "  -1354.8558898930407,\n",
       "  -1362.6124249516154,\n",
       "  -1370.4467072911843,\n",
       "  -1378.3549151702568,\n",
       "  -1386.333597751424,\n",
       "  -1394.3796278957348,\n",
       "  -1402.4901623514634,\n",
       "  -1410.6626079784735,\n",
       "  -1418.8945929333481,\n",
       "  -1427.1839419582207,\n",
       "  -1435.5286550847457,\n",
       "  -1443.9268891961647,\n",
       "  -1452.376941993818,\n",
       "  -1460.8772379964253,\n",
       "  -1469.4263162658208,\n",
       "  -1478.0228196053654],\n",
       " 0.015,\n",
       " 0.00738069299475812,\n",
       " 0.022619307005241878)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapCon_chrom(id, ch='X', \n",
    "            path_targets=f\"{basePath}/{id}/{id}.hdf5\",\n",
    "            h5_path1000g='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "            meta_path_ref='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv', \n",
    "            folder_out=\"./test_output/contaminate/first_try/\", prefix_out=\"fast_\",\n",
    "            e_model=\"readcount_contam\", p_model=\"SardHDF5\", post_model=\"Standard\", save=True, save_fp=False, \n",
    "            n_ref=2504, diploid_ref=True, exclude_pops=[], conPop=[], readcounts=True, random_allele=False,\n",
    "            c=np.arange(0, 0.2, 0.005), roh_in=1, roh_out=0, roh_jump=300, e_rate=0.002, e_rate_ref=1e-3, \n",
    "            cutoff_post = 0.999, max_gap=0, roh_min_l = 0.01, logfile=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: I10939\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 1 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/iberian_BAM/I10939/I10939.hdf5\n",
      "\n",
      "Loaded 47898 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from /mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5\n",
      "\n",
      "Intersection on Positions: 47895\n",
      "Nr of Matching Refs: 47895 / 47895\n",
      "Ref/Alt Allele Matching: 47895 / 47895\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 47895 / 47895\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 47895)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 47895)\n",
      "Extraction of 3775 Haplotypes complete\n",
      "Subset to markers with data: 6857 / 47895\n",
      "Fraction SNPs covered: 0.1432\n",
      "Successfully saved target individual data to: ./test_output/contaminate/first_try/I10939/chrX/fast_\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.21613\n",
      "Successfully loaded Data from: ./test_output/contaminate/first_try/I10939/chrX/fast_\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tI10939\n",
      "ch\tX\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "conPop\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t/mnt/archgen/users/yilei/Data/iberian_BAM/I10939/I10939.hdf5\n",
      "h5_path1000g\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv\n",
      "folder_out\t./test_output/contaminate/first_try/\n",
      "prefix_out\tfast_\n",
      "c\t0.025\n",
      "roh_in\t1\n",
      "roh_out\t0\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.001\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "##### args: [0.025 0.01 ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.02500001 0.01      ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.025      0.01000001] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0. 0.] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.e-08 0.e+00] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.e+00 1.e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01228479 0.00491392] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.0122848  0.00491392] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01228479 0.00491393] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00231356 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00231357 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [2.31356061e-03 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00802195 0.00281315] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00802196 0.00281315] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00802195 0.00281316] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00918483 0.00132581] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00918484 0.00132581] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.00918483 0.00132582] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01426228 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01426229 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.42622769e-02 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01976167 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01976168 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.97616692e-02 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01876252 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01876253 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.87625153e-02 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01848557 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01848558 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.84855708e-02 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01850168 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01850169 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.8501675e-02 1.0000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01850147 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01850148 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.85014736e-02 1.00000000e-08] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "----------- res:       fun: 1227.4420364024022\n",
      " hess_inv: <2x2 LbfgsInvHessProduct with dtype=float64>\n",
      "      jac: array([-4.54747351e-05,  7.43042665e+02])\n",
      "  message: b'CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH'\n",
      "     nfev: 36\n",
      "      nit: 9\n",
      "   status: 0\n",
      "  success: True\n",
      "        x: array([0.01850147, 0.        ]) --------------\n",
      "##### args: [0.01850147 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01870147 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [0.01830147 0.        ] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [1.86014736e-02 1.00000000e-04] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Reference Number: 3775\n",
      "##### args: [ 1.86014736e-02 -1.00000000e-04] #####\n",
      "Minimum Genetic Map: 0.2090 Morgan\n",
      "Maximum Genetic Map: 2.0205 Morgan\n",
      "Gaps bigger than 0.1 cM: 494\n",
      "Maximum Gap: 1.3339 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "cannot estimate the Hessian of the loglikelihood around [0.01850147 0.        ]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([0.01850147, 0.        ]), array([nan, nan]))"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hapCon_chrom_2d(id, ch='X', \n",
    "            path_targets=f\"{basePath}/{id}/{id}.hdf5\",\n",
    "            h5_path1000g='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "            meta_path_ref='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv', \n",
    "            folder_out=\"./test_output/contaminate/first_try/\", prefix_out=\"fast_\",\n",
    "            e_model=\"readcount_contam\", p_model=\"SardHDF5\", post_model=\"Standard\", save=True, save_fp=False, \n",
    "            n_ref=2504, diploid_ref=True, exclude_pops=[], conPop=[], readcounts=True, random_allele=False,\n",
    "            c=0.025, roh_in=1, roh_out=0, roh_jump=300, e_rate=0.01, e_rate_ref=1e-3, \n",
    "            cutoff_post = 0.999, max_gap=0, roh_min_l = 0.01, logfile=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([17343, 17344]),)\n"
     ]
    }
   ],
   "source": [
    "f = h5py.File('/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chrX.hdf5', 'r')\n",
    "pos = f['variants/POS']\n",
    "rec = f['variants/MAP']\n",
    "refs = f['variants/REF'][:]\n",
    "alts = f['variants/ALT'][:, 0]\n",
    "gts = np.array(f[\"calldata/GT\"])\n",
    "nloci, nind, _ = gts.shape\n",
    "gts = gts.reshape(nloci, nind*2)\n",
    "# exclude haplotypes with missing genotypes (for male x chromosome, one column is always -1, for example)\n",
    "gts = gts.astype(np.float32)\n",
    "gts[gts == -1] = np.nan\n",
    "freqs = np.nanmean(gts, axis=1)\n",
    "assert(len(pos) == len(refs))\n",
    "assert(len(pos) == len(alts))\n",
    "assert(len(pos) == len(freqs))\n",
    "\n",
    "# write a bed file for the focal site, note that bed file uses 0-based coordinate\n",
    "pos = np.array(pos)\n",
    "print(np.where(pos == 47635874))\n",
    "\n",
    "\n",
    "# pos = pos-1\n",
    "# with open('/mnt/archgen/users/yilei/bin/angsd/RES/1240kChrX.bed', 'w') as out:\n",
    "#     for p in pos:\n",
    "#         out.write(f'X\\t{p}\\t{1+p}\\n')\n",
    "\n",
    "\n",
    "\n",
    "# with open('/mnt/archgen/users/yilei/bin/angsd/RES/1240kChrX.txt','w') as out:\n",
    "#     for p, ref, freq, alt in zip(pos, refs, freqs, alts):\n",
    "#         out.write(f'{p}\\t{ref}\\t{freq}\\t+\\t{alt}\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom_BFGS\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom_2d\n",
    "\n",
    "mle, low95, up95 = hapCon_chrom_BFGS(\"I0843\", 'X', save=False, save_fp=False, n_ref=2504, diploid_ref=True, \n",
    "        exclude_pops=[], conPop=[], e_model=\"readcount_contam\", p_model=\"SardHDF5\", \n",
    "        readcounts=True, random_allele=False, post_model=\"Standard\", \n",
    "        path_targets = '/mnt/archgen/users/yilei/Data/iberian_BAM/HDF5/I0843_1240k.hdf5', \n",
    "        folder_out='/mnt/archgen/users/yilei/Data/iberian_BAM/hapCon/',\n",
    "        h5_path1000g='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "        meta_path_ref='/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/meta_df_all.csv', \n",
    "        prefix_out=\"\", c=0.025, roh_in=1, roh_out=0, roh_jump=300, e_rate=0.0019, e_rate_ref=0.0,\n",
    "        max_gap=0, cutoff_post = 0.999, roh_min_l = 0.01, logfile=False)\n",
    "print(f'BFGS estimate: {mle}({low95} - {up95})')"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
