{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "Simulation autosomes with (fake) readcount data. Just a trial run to establish pipeline"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os as os\n",
    "import sys as sys\n",
    "import multiprocessing as mp\n",
    "\n",
    "print(f\"CPU Count: {mp.cpu_count()}\")\n",
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/create1000G_Mosaic\")\n",
    "from createMosaicsMulti import Mosaic_1000G_Multi\n",
    "from createMosaicsMulti import create_individual_mosaic\n",
    "\n",
    "path = \"/mnt/archgen/users/yilei/tools/hapROH\"   # The Path on Yilei's remote space\n",
    "os.chdir(path)  # Set the right Path (in line with Atom default)\n",
    "print(f\"Set path to: {os.getcwd()}\") # Show the current working directory."
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "source": [
    "# actual simulation\n",
    "#### Create the parameters array for the starmap:\n",
    "base_path=\"./simulated/1000G_Mosaic/TSI/\" \n",
    "path1000G=\"/mnt/archgen/users/yilei/Data/1000G/1000g1240khdf5/all1240/chr\"\n",
    "pop_list=[\"TSI\"] \n",
    "n=5   # Number of Individuals\n",
    "ch=1\n",
    "chunk_length=0.0025\n",
    "l = 12.5  # length of simulated ROH, in centiMorgen\n",
    "n_blocks=10 # How many blocks will be copied in\n",
    "\n",
    "# parameters for faking readcount data\n",
    "cov = 1.0\n",
    "con = 0.05\n",
    "err_rate=1e-3\n",
    "\n",
    "\n",
    "create_individual_mosaic(base_path, path1000G, pop_list, n, ch, chunk_length, l, n_blocks, cov, con, err_rate)\n",
    "\n",
    "\n",
    "### Create the List of Parameter Lists (input for starmap)\n",
    "# prms = []\n",
    "\n",
    "# for l in [2, 4, 6, 8, 10]:\n",
    "#     new_par = [base_path, path1000G, pop_list, n, ch, chunk_length, l, n_blocks]\n",
    "#     prms.append(new_par)  # Append to the Parameters\n",
    "\n",
    "# assert(len(prms[0])==8)   # The function takes 8 Parameters as input\n",
    "# multi_run(create_individual_mosaic, prms, processes = 2)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "First, run simulated samples as target in hapROH without assuming contamination "
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "sys.path.insert(0, \"/mnt/archgen/users/yilei/tools/hapROH/package\")  # hack to get local package first in path [FROM HARALD - DELETE!!!]\n",
    "from hapsburg.PackagesSupport.hapsburg_run import hapsb_ind  # Need this import"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "# readcount mode\n",
    "for i in range(5):\n",
    "    hapsb_ind(iid=f\"iid{i}\", chs=range(3,4), \n",
    "          path_targets='./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5',\n",
    "          h5_path1000g='../..//Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='../../Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./simulated/1000G_Mosaic/TSI/chr3_5cM', prefix_out='rc', \n",
    "          e_model='readcount', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=True, random_allele=False, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "# diploid mode\n",
    "for i in range(5):\n",
    "    hapsb_ind(iid=f\"iid{i}\", chs=range(3,4), \n",
    "          path_targets='./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5',\n",
    "          h5_path1000g='../..//Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='../../Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./simulated/1000G_Mosaic/TSI/chr3_5cM', prefix_out='dip', \n",
    "          e_model='diploid_gt', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=False, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "# haploid mode\n",
    "for i in range(5):\n",
    "    hapsb_ind(iid=f\"iid{i}\", chs=range(3,4), \n",
    "          path_targets='./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5',\n",
    "          h5_path1000g='../..//Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='../../Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./simulated/1000G_Mosaic/TSI/chr3_5cM', prefix_out='hap', \n",
    "          e_model='haploid', p_model='SardHDF5', \n",
    "          post_model='Standard', processes=1, delete=False, output=True, save=True, \n",
    "          save_fp=False, n_ref=2504, exclude_pops=[], readcounts=False, random_allele=True, \n",
    "          roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0, \n",
    "          cutoff_post=0.999, max_gap=0, roh_min_l=0.01, \n",
    "          logfile=False, combine=False, file_result='_roh_full.csv')"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now run hapROH with contamination"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "from hapsburg.PackagesSupport.hapsburg_run import hapCon_chrom"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "for i in range(5):\n",
    "    hapCon_chrom(iid=f\"iid{i}\", ch=3, save=True, save_fp=False, n_ref=2504, diploid_ref=True, exclude_pops=[], \n",
    "                e_model=\"readcount_contam\", p_model=\"SardHDF5\", readcounts=True,random_allele=False, post_model=\"Standard\", \n",
    "                path_targets='./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5',\n",
    "          h5_path1000g='../..//Data/1000G/1000g1240khdf5/all1240/chr',\n",
    "          meta_path_ref='../../Data/1000G/meta_df_all.csv', \n",
    "          folder_out='./simulated/1000G_Mosaic/TSI/chr3_5cM', prefix_out=\"rcCon\",\n",
    "                c=0.05, roh_in=1, roh_out=20, roh_jump=300, e_rate=0.01, e_rate_ref=0.0,\n",
    "                max_gap=0, cutoff_post = 0.999, roh_min_l = 0.01, logfile=False)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: iid0\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 5 individuals\n",
      "HDF5 loaded from ./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ../..//Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77652\n",
      "Nr of Matching Refs: 77652 / 77652\n",
      "Ref/Alt Allele Matching: 77652 / 77652\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 77652 / 77652\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77652)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77652)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 48830 / 77652\n",
      "Fraction SNPs covered: 0.6288\n",
      "Successfully saved target individual data to: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid0/chr3/rcCon\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.57678\n",
      "Successfully loaded Data from: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid0/chr3/rcCon\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tiid0\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "h5_path1000g\t../..//Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t../../Data/1000G/meta_df_all.csv\n",
      "folder_out\t./simulated/1000G_Mosaic/TSI/chr3_5cM\n",
      "prefix_out\trcCon\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 297\n",
      "Maximum Gap: 0.3910 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 48830)\n",
      "Loaded Observations:\n",
      "(2, 48830)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2399.41632 mB\n",
      "Total Log likelihood: -23183.676\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid0/chr3/rcCon.\n",
      "Successfully loaded for PP. from ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid0/chr3/rcCon\n",
      "Fraction Markers in ROH: 0.3085\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=7 ROH Blocks > 1.0 cM\n",
      "Longest Block: 5.42 cM\n",
      "Successfully saved to ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid0/chr3/rcCon/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: iid1\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 5 individuals\n",
      "HDF5 loaded from ./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ../..//Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77652\n",
      "Nr of Matching Refs: 77652 / 77652\n",
      "Ref/Alt Allele Matching: 77652 / 77652\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 77652 / 77652\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77652)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77652)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 49035 / 77652\n",
      "Fraction SNPs covered: 0.6315\n",
      "Successfully saved target individual data to: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid1/chr3/rcCon\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.58473\n",
      "Successfully loaded Data from: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid1/chr3/rcCon\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tiid1\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "h5_path1000g\t../..//Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t../../Data/1000G/meta_df_all.csv\n",
      "folder_out\t./simulated/1000G_Mosaic/TSI/chr3_5cM\n",
      "prefix_out\trcCon\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 302\n",
      "Maximum Gap: 0.3985 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 49035)\n",
      "Loaded Observations:\n",
      "(2, 49035)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2415.34976 mB\n",
      "Total Log likelihood: -23715.979\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid1/chr3/rcCon.\n",
      "Successfully loaded for PP. from ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid1/chr3/rcCon\n",
      "Fraction Markers in ROH: 0.2969\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=10 ROH Blocks > 1.0 cM\n",
      "Longest Block: 5.08 cM\n",
      "Successfully saved to ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid1/chr3/rcCon/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: iid2\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 5 individuals\n",
      "HDF5 loaded from ./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ../..//Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77652\n",
      "Nr of Matching Refs: 77652 / 77652\n",
      "Ref/Alt Allele Matching: 77652 / 77652\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 77652 / 77652\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77652)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77652)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 49264 / 77652\n",
      "Fraction SNPs covered: 0.6344\n",
      "Successfully saved target individual data to: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid2/chr3/rcCon\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.58201\n",
      "Successfully loaded Data from: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid2/chr3/rcCon\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tiid2\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "h5_path1000g\t../..//Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t../../Data/1000G/meta_df_all.csv\n",
      "folder_out\t./simulated/1000G_Mosaic/TSI/chr3_5cM\n",
      "prefix_out\trcCon\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2325 Morgan\n",
      "Gaps bigger than 0.1 cM: 304\n",
      "Maximum Gap: 0.2593 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 49264)\n",
      "Loaded Observations:\n",
      "(2, 49264)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2426.032128 mB\n",
      "Total Log likelihood: -22748.572\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid2/chr3/rcCon.\n",
      "Successfully loaded for PP. from ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid2/chr3/rcCon\n",
      "Fraction Markers in ROH: 0.3283\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=8 ROH Blocks > 1.0 cM\n",
      "Longest Block: 5.24 cM\n",
      "Successfully saved to ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid2/chr3/rcCon/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: iid3\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 5 individuals\n",
      "HDF5 loaded from ./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ../..//Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77652\n",
      "Nr of Matching Refs: 77652 / 77652\n",
      "Ref/Alt Allele Matching: 77652 / 77652\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 77652 / 77652\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77652)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77652)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 49142 / 77652\n",
      "Fraction SNPs covered: 0.6328\n",
      "Successfully saved target individual data to: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid3/chr3/rcCon\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.58740\n",
      "Successfully loaded Data from: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid3/chr3/rcCon\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tiid3\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "h5_path1000g\t../..//Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t../../Data/1000G/meta_df_all.csv\n",
      "folder_out\t./simulated/1000G_Mosaic/TSI/chr3_5cM\n",
      "prefix_out\trcCon\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 293\n",
      "Maximum Gap: 0.4228 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 49142)\n",
      "Loaded Observations:\n",
      "(2, 49142)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2420.789248 mB\n",
      "Total Log likelihood: -23505.095\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid3/chr3/rcCon.\n",
      "Successfully loaded for PP. from ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid3/chr3/rcCon\n",
      "Fraction Markers in ROH: 0.2801\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=8 ROH Blocks > 1.0 cM\n",
      "Longest Block: 5.09 cM\n",
      "Successfully saved to ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid3/chr3/rcCon/roh.csv\n",
      "Using Rescaled HMM.\n",
      "Loaded Pre Processing Model: SardHDF5\n",
      "Loading Individual: iid4\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 5 individuals\n",
      "HDF5 loaded from ./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "\n",
      "Loaded 77652 variants\n",
      "Loaded 2504 individuals\n",
      "HDF5 loaded from ../..//Data/1000G/1000g1240khdf5/all1240/chr3.hdf5\n",
      "\n",
      "Intersection on Positions: 77652\n",
      "Nr of Matching Refs: 77652 / 77652\n",
      "Ref/Alt Allele Matching: 77652 / 77652\n",
      "Flipped Ref/Alt Alleles for 0 SNPs\n",
      "Together: 77652 / 77652\n",
      "2504 / 2504 Individuals included in Reference\n",
      "Extracting up to 2504 Individuals\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (2, 77652)\n",
      "Extraction of 2 Haplotypes complete\n",
      "Flipping Ref/Alt Allele in target for 0 SNPs...\n",
      "Exctraction of hdf5 done. Subsetting...!\n",
      "size of gts matrix: (5008, 77652)\n",
      "Extraction of 5008 Haplotypes complete\n",
      "Subset to markers with data: 49066 / 77652\n",
      "Fraction SNPs covered: 0.6319\n",
      "Successfully saved target individual data to: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid4/chr3/rcCon\n",
      "Loading Readcounts...\n",
      "Mean Readcount on markers with data: 1.57690\n",
      "Successfully loaded Data from: ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid4/chr3/rcCon\n",
      "Loaded Emission Model: readcount_contam\n",
      "Loaded Transition Model: model\n",
      "Loaded Post Processing Model: Standard\n",
      "\n",
      "Parameters in hapCon_chrom:\n",
      "iid\tiid4\n",
      "ch\t3\n",
      "save\tTrue\n",
      "save_fp\tFalse\n",
      "n_ref\t2504\n",
      "diploid_ref\tTrue\n",
      "exclude_pops\t[]\n",
      "e_model\treadcount_contam\n",
      "p_model\tSardHDF5\n",
      "readcounts\tTrue\n",
      "random_allele\tFalse\n",
      "post_model\tStandard\n",
      "path_targets\t./simulated/1000G_Mosaic/TSI/chr3_5cM/data.h5\n",
      "h5_path1000g\t../..//Data/1000G/1000g1240khdf5/all1240/chr\n",
      "meta_path_ref\t../../Data/1000G/meta_df_all.csv\n",
      "folder_out\t./simulated/1000G_Mosaic/TSI/chr3_5cM\n",
      "prefix_out\trcCon\n",
      "c\t0.05\n",
      "roh_in\t1\n",
      "roh_out\t20\n",
      "roh_jump\t300\n",
      "e_rate\t0.01\n",
      "e_rate_ref\t0.0\n",
      "max_gap\t0\n",
      "cutoff_post\t0.999\n",
      "roh_min_l\t0.01\n",
      "logfile\tFalse\n",
      "\n",
      "\n",
      "Minimum Genetic Map: 0.0000 Morgan\n",
      "Maximum Genetic Map: 2.2326 Morgan\n",
      "Gaps bigger than 0.1 cM: 311\n",
      "Maximum Gap: 0.2793 cM\n",
      "Upper Gap Cutoff: 5.0000 cM\n",
      "Loaded Transition and Emission Matrix:\n",
      "(3, 3)\n",
      "(5009, 49066)\n",
      "Loaded Observations:\n",
      "(2, 49066)\n",
      "Reference Number: 5008\n",
      "Memory Usage at end of HMM:\n",
      "Memory Usage: 2417.606656 mB\n",
      "Total Log likelihood: -22839.957\n",
      "Finished Calculation State Posteriors\n",
      "Saved Zero State Posterior to folder ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid4/chr3/rcCon.\n",
      "Successfully loaded for PP. from ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid4/chr3/rcCon\n",
      "Fraction Markers in ROH: 0.2986\n",
      "Merged n=0 gaps < 0 M\n",
      "Called n=5 ROH Blocks > 1.0 cM\n",
      "Longest Block: 5.41 cM\n",
      "Successfully saved to ./simulated/1000G_Mosaic/TSI/chr3_5cM/iid4/chr3/rcCon/roh.csv\n"
     ]
    }
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.8.10",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.8.10 64-bit"
  },
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}